{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(Seurat)\n",
    "#library(SeuratData)\n",
    "library(ggplot2)\n",
    "source(\"../src/process_pdac.R\")\n",
    "source(\"../src/pdac_plots.R\")\n",
    "#library(parallelDist)\n",
    "library(grid)\n",
    "library(dplyr)\n",
    "library(pheatmap)\n",
    "library(DoubletFinder)\n",
    "library(ggpubr)\n",
    "library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(BSgenome.Hsapiens.UCSC.hg19)\n",
    "library(robustbase)\n",
    "library(gridExtra)\n",
    "library(RColorBrewer)\n",
    "library(stringr)\n",
    "library(monocle)\n",
    "library(fgsea)\n",
    "library(GEOquery)\n",
    "source(\"../src/motif_analysis.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PDAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path <- file.path(\"..\",\"data\")\n",
    "pdac_matrix_file_name <- \"CRA001160.matrix\"\n",
    "pdac_anno_file_name <- \"CRA001160.celltypes.tsv\"\n",
    "pdac_matrix_path <- file.path( base_path, pdac_matrix_file_name )\n",
    "pdac_annotation_path <- file.path( base_path, pdac_anno_file_name )\n",
    "\n",
    "pdac_anno_dt <- process_annotations( pdac_annotation_path )\n",
    "seurat_obj <- create_full_seurat_object( read_gene_exp_mat( \n",
    "    pdac_matrix_path ), pdac_anno_dt )\n",
    "meta_data_dt <- data.table( seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "\n",
    "doublet_cells <- fread( file.path( base_path, \"cell_doublets.tsv\" ),header=F) %>% .$V1\n",
    "\n",
    "filtered_cells <- meta_data_dt[!cell.name %in% doublet_cells,cell.name]\n",
    "pdac_anno_dt <- pdac_anno_dt[cell.name %in% filtered_cells,]\n",
    "\n",
    "seurat_obj <- subset( seurat_obj, cells=filtered_cells )\n",
    "seurat_obj <- NormalizeData( seurat_obj )\n",
    "meta_data_dt <- data.table( seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "edge_info_all <- readRDS( file.path( base_path, \"edge_info_all_doublet_filtered_celltypes.rds\" ) )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison with trajectory analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_obj <- readRDS(file.path(base_path,\"ductal_malignant_cds_obj.rds\"))\n",
    "cell_type <- \"Ductal cell type 1\"\n",
    "cds_meta_data_dt <- data.table( pData(cds_obj) )\n",
    "cells <- cds_meta_data_dt[!cell.name %in% doublet_cells,cell.name]\n",
    "# cds_obj <- cds_obj[,cds_meta_data_dt$cell.name]\n",
    "cds_meta_data_dt$sub_cluster <- cds_meta_data_dt$cluster\n",
    "cds_meta_data_dt[cell.name %in% edge_info_all$edge_center_dt[normal_cell_type==cell_type & cell_category == \"edge\",cell.name],sub_cluster:=paste(\"Edge\", cell_type)]\n",
    "cds_meta_data_dt[!grepl(\"Edge\",sub_cluster) & cluster == cell_type,sub_cluster:=paste(\"Non-Edge\", cell_type)]\n",
    "df <- as.data.frame(cds_meta_data_dt[,!c(\"cell.name\")])\n",
    "rownames(df) <- cds_meta_data_dt$cell.name\n",
    "pData(cds_obj) <- df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=14,repr.plot.height=6)\n",
    "p1 <- plot_cell_trajectory( cds_obj, color_by=\"sub_cluster\" ) + labs(color=\"Cell Type\") + theme_classic(base_size=15) +\n",
    "theme(legend.position=c(1,0.8),legend.justification = \"right\",legend.text=element_text(size=15)) +\n",
    "scale_color_discrete(name=\"Cell Type\",labels=c(\"Ductal cell type 2\"=\"Malignant Ductal\",\n",
    "                                              \"Edge Ductal cell type 1\"=\"Outlier Ductal\",\n",
    "                                              \"Non-Edge Ductal cell type 1\"=\"Non-Outlier Ductal\"))\n",
    "\n",
    "p2 <- ggboxplot( cds_meta_data_dt[sub_cluster!=\"Acinar cell\",], x=\"State\",y=\"Pseudotime\",color=\"sub_cluster\" ) + \n",
    "theme_classic(base_size=15) + xlab(\"State\") +\n",
    "stat_compare_means(comparisons=list(c(\"Non-Edge Acinar cell\",\"Edge Acinar cell\")),label.y=7) +\n",
    "theme(legend.position=\"none\")\n",
    "final <- ggarrange(p1,p2,nrow=1)\n",
    "print(final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=14,repr.plot.height=6)\n",
    "p1 <- plot_cell_trajectory( cds_obj, color_by=\"sub_cluster\" ) + labs(color=\"\") + theme_classic(base_size=15) +\n",
    "theme(legend.position=c(1,0.8),legend.justification = \"right\",legend.text=element_text(size=15)) #+ \n",
    "# scale_color_manual(breaks=c(\"Ductal cell type 2\"=\"Malignant Ductal\",\"Edge Ductal Cell Type 1\"=\"Outlier Ductal\",\n",
    "#                            \"Non-Edge Ductal cell type 1\"=\"Non-Outlier Ductal\"))\n",
    "\n",
    "\n",
    "p2 <- ggboxplot( cds_meta_data_dt[sub_cluster!=cell_type,], x=\"sub_cluster\",y=\"Pseudotime\" ) + \n",
    "theme_classic(base_size=15) + xlab(\"\") +\n",
    "stat_compare_means(comparisons=list(c(\"Non-Edge Ductal cell type 1\",\"Edge Ductal cell type 1\")),label.y=20)\n",
    "final <- ggarrange(p1,p2,nrow=1)\n",
    "print(final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10,repr.plot.height=6)\n",
    "# p1 <- plot_cell_trajectory( cds_obj, color_by=\"sub_cluster\" ) + labs(color=\"\") + theme_classic(base_size=15) + \n",
    "# facet_grid(rows=vars(sub_cluster),cols=vars(State)) + theme(legend.position=\"bottom\",legend.justification = \"right\",legend.text=element_text(size=15)) #+ \n",
    "# # scale_color_manual(breaks=c(\"Ductal cell type 2\"=\"Malignant Ductal\",\"Edge Ductal Cell Type 1\"=\"Outlier Ductal\",\n",
    "# #                            \"Non-Edge Ductal cell type 1\"=\"Non-Outlier Ductal\"))\n",
    "\n",
    "\n",
    "p2 <- ggboxplot( cds_meta_data_dt, color=\"sub_cluster\",y=\"Pseudotime\",x=\"State\" ) + \n",
    "theme_classic(base_size=20) + labs(color=\"Cell Type\") +\n",
    "stat_compare_means(comparisons=list(c(\"Non-Edge Ductal cell type 1\",\"Edge Ductal cell type 1\"))) + \n",
    "theme(legend.position=\"bottom\",legend.text=element_text(size=15))\n",
    "p2\n",
    "# final <- ggarrange(p1,p2,ncol=1)\n",
    "# print(final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sub-sampling reads and genes from edge cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"process_pdac.R\")\n",
    "source(\"pdac_plots.R\")\n",
    "norm_type <- \"Acinar cell\"\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=3)\n",
    "#theme_set(theme_classic(base_size = 15))\n",
    "resample_info_list <- resample_edge_cells( seurat_obj, pdac_anno_dt, edge_info_all$edge_center_dt, c(\"Acinar cell\"), c(\"Ductal cell type 2\"), \"reads\",\n",
    "                                      no_tumour_adjacent=F)\n",
    "\n",
    "#Before resampling\n",
    "p1 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes(x=cell_category_before_resampling,y=log(nCount_RNA_before_resampling))) + \n",
    "stat_compare_means(aes(x=cell_category_before_resampling,y=log(nCount_RNA_before_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() +\n",
    "xlab(\"Acinar cell category\") + ylab(\"Log(Library size)\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\n",
    "                                                                                    \"edge\"=\"Edge\"))\n",
    "\n",
    "#After resampling\n",
    "p2 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes(x=cell_category_after_resampling,y=log(nCount_RNA_after_resampling))) +\n",
    "stat_compare_means(aes(x=cell_category_after_resampling,y=log(nCount_RNA_after_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() + ylab(\"\") +\n",
    " xlab(\"Acinar cell category\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\"edge\"=\"Edge\"))\n",
    " \n",
    "p3 <- plot_skewness(resample_info_list$edge_info$control_dist_dt) + theme_classic()\n",
    "\n",
    "p4 <- plot_edge_distance_ratio(resample_info_list$edge_info$edge_malignant_dist_dt) + theme_classic()\n",
    "                              \n",
    "final <- ggarrange(p1,p2,p3,p4,nrow=1,ncol=4,labels=\"auto\")\n",
    "print(final)\n",
    "\n",
    "num_retained_edge_cells <- nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\" & \n",
    "                                         cell_category_after_resampling == \"edge\",] )\n",
    "print(paste(\"# edge cells =\",nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\",])))\n",
    "print(paste(\"# common edge cells before and after resampling =\",num_retained_edge_cells))\n",
    "ggsave(\"supp_library_resampling.png\", final, height=3, width=8 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"process_pdac.R\")\n",
    "source(\"pdac_plots.R\")\n",
    "norm_type <- \"Acinar cell\"\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=3)\n",
    "#theme_set(theme_classic(base_size = 15))\n",
    "resample_info_list <- resample_edge_cells( seurat_obj, pdac_anno_dt, edge_info_all$edge_center_dt, c(\"Acinar cell\"), \n",
    "                                           c(\"Ductal cell type 2\"), \"genes\", no_tumour_adjacent=F)\n",
    "\n",
    "#Before resampling\n",
    "p1 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes(x=cell_category_before_resampling,y=log(nFeature_RNA_before_resampling))) + \n",
    "stat_compare_means(aes(x=cell_category_before_resampling,y=log(nFeature_RNA_before_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() +\n",
    "xlab(\"Acinar cell category\") + ylab(\"Log(Library size)\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\n",
    "                                                                                    \"edge\"=\"Edge\"))\n",
    "\n",
    "#After resampling\n",
    "p2 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes (x=cell_category_after_resampling,y=log(nFeature_RNA_after_resampling))) +\n",
    "stat_compare_means(aes(x=cell_category_after_resampling,y=log(nFeature_RNA_after_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() + ylab(\"\") +\n",
    "xlab(\"Acinar cell category\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\"edge\"=\"Edge\"))\n",
    " \n",
    "p3 <- plot_skewness(resample_info_list$edge_info$control_dist_dt) + theme_classic()\n",
    "\n",
    "p4 <- plot_edge_distance_ratio(resample_info_list$edge_info$edge_malignant_dist_dt) + theme_classic()\n",
    "                              \n",
    "final <- ggarrange(p1,p2,p3,p4,nrow=1,ncol=4,labels=\"auto\")\n",
    "print(final)\n",
    "\n",
    "num_retained_edge_cells <- nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\" & \n",
    "                                         cell_category_after_resampling == \"edge\",] )\n",
    "print(paste(\"# edge cells =\",nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\",])))\n",
    "print(paste(\"# common edge cells before and after resampling =\",num_retained_edge_cells))\n",
    "ggsave(\"supp_genes_resampling.png\", final, height=3, width=8 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Justifying merging normal and tumour-adjacent acinar cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_type <- \"Acinar cell\"\n",
    "tumour_adjacent_cells <- meta_data_dt[cluster == norm_type & sample_type == \"tumour\",cell.name]\n",
    "\n",
    "# acinar_edge_info <- add_edge_center_annotation( seurat_obj, pdac_anno_dt, \n",
    "#                                                       malignant_cell_types=c(\"Ductal cell type 2\"), \n",
    "#                                   normal_cell_types=c(norm_type), num_pcs=num_pcs, \n",
    "#                                   pairwise_pca=T, perform_control=F, no_tumour_adjacent=T )\n",
    "\n",
    "cell_type_acinar_edge_center_dt <- acinar_edge_info$edge_center_dt\n",
    "diff_exp_dt <- get_edge_de_genes_dt( seurat_obj, cell_type_acinar_edge_center_dt, \n",
    "                                    tumour_adjacent_cells = tumour_adjacent_cells )\n",
    "diff_exp_dt <- diff_exp_dt[p_val_adj < 0.01,]\n",
    "\n",
    "acinar_edge_cells <- cell_type_acinar_edge_center_dt[cell_category == \"edge\",cell.name]\n",
    "non_acinar_edge_cells <- cell_type_acinar_edge_center_dt[cell_category != \"edge\",cell.name]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=3)\n",
    "theme_set(theme_classic())\n",
    "\n",
    "merged_diff_exp_dt <- merge( diff_exp_dt[de_set == \"tumour_adjacent_vs_center\"],\n",
    "                         diff_exp_dt[de_set == \"edge_vs_center\"], by=\"gene_name\",\n",
    "                           suffixes=c(\".adj_vs_center\",\".edge_vs_center\"))\n",
    "p1 <- ggplot( merged_diff_exp_dt ) + geom_point(aes(x=avg_logFC.edge_vs_center,y=avg_logFC.adj_vs_center)) + \n",
    "geom_line(aes(x=avg_logFC.edge_vs_center,y=avg_logFC.edge_vs_center),color=\"gray\") + \n",
    "geom_hline(aes(yintercept=0)) + geom_vline(aes(xintercept=0)) + xlab(\"Average Log FC\\n(Edge vs Non-Edge)\") +\n",
    "ylab(\"Average Log FC\\n(Tumor-adjacent vs Non-Edge)\")\n",
    "print(paste(\"R^2\"=cor.test(merged_diff_exp_dt$avg_logFC.edge_vs_center,merged_diff_exp_dt$avg_logFC.adj_vs_center)))\n",
    "\n",
    "plot_dt <- melt( merged_diff_exp_dt[,.(gene_name,avg_logFC.edge_vs_center,avg_logFC.adj_vs_center)],\n",
    "                id.vars=\"gene_name\", value.name=\"Average Log FC\")\n",
    "p2 <- ggplot( plot_dt ) + geom_point(aes(x=variable,y=`Average Log FC`),position=\"jitter\",color=\"gray\") +\n",
    "geom_boxplot(aes(x=variable,y=`Average Log FC`),outlier.shape=NA,fill=NA,size=1.0) + \n",
    "scale_x_discrete(\"\",labels=c(\"avg_logFC.edge_vs_center\" = \"Edge vs\\n Non-Edge\",\n",
    "                          \"avg_logFC.adj_vs_center\" = \"Tumor-Adjacent vs\\n Non-Edge\")) + \n",
    "stat_compare_means(aes(x=variable,y=`Average Log FC`),label.x=1.5,label=\"p.signif\")\n",
    "\n",
    "acinar_edge_center_dt <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\",.(cell.name,dist_from_normal_medoid,dist_from_malignant_medoid)]\n",
    "acinar_edge_center_dt <- merge( acinar_edge_center_dt, meta_data_dt[,.(cell.name,sample_type)] )\n",
    "melted_dt <- melt(acinar_edge_center_dt[sample_type==\"tumour\",][,!c(\"sample_type\")],id.vars=c(\"cell.name\"))\n",
    "p3 <- ggplot( melted_dt ) + stat_compare_means(aes(x=variable,y=value),label=\"..p.format..\",method.args=list(\"alternative\"=\"g\"),label.x=1.5) +\n",
    " geom_boxplot(aes(x=variable,y=value)) + scale_x_discrete(c(\"variable\"=\"\"), labels=c(\"dist_from_normal_medoid\"=\"From\\nacinar\",\n",
    "                                                           \"dist_from_malignant_medoid\"=\"From\\nmalignant ductal\")) + ylab(\"Distance in PC space\")\n",
    "ggarrange(p1,p2,p3ncol=3,nrow=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(paste0(\"# of DE genes (edge vs center) \", nrow(diff_exp_dt[de_set == \"edge_vs_center\"])))\n",
    "print(paste0(\"# of DE genes (tumour-adjacent vs center) \", nrow(diff_exp_dt[de_set == \"tumour_adjacent_vs_center\"])))\n",
    "print(paste0(\"# of common DE genes \", nrow(merged_diff_exp_dt)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Edge-ness fingerprint plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cell_types <- unique(pdac_anno_dt$cluster)\n",
    "across_refs_edge_center_dt <- data.table()\n",
    "num_pcs = 50\n",
    "\n",
    "malignant_cell_types <- c(\"Ductal cell type 2\")\n",
    "normal_cell_types <- setdiff( unique(pdac_anno_dt$cluster), malignant_cell_types)\n",
    "edge_info_all <- add_edge_center_annotation( seurat_obj, pdac_anno_dt, malignant_cell_types=malignant_cell_types, \n",
    "                                  normal_cell_types=normal_cell_types, num_pcs=num_pcs, \n",
    "                                  pairwise_pca=T, perform_control=T, no_tumour_adjacent=F )\n",
    "\n",
    "saveRDS(edge_info_all,\"edge_info_all_doublet_filtered_celltypes.rds\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Checking which sample edge cells predominantly come from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_info_all_merged <- merge( edge_info_all$edge_center_dt, meta_data_dt[,.(cell.name,sample)] )\n",
    "ordered_samples <- meta_data_dt[,.N,by=sample][order(-N)]$sample\n",
    "edge_info_all_merged$sample <- factor( edge_info_all_merged$sample, levels = ordered_samples)\n",
    "\n",
    "cell_types_present <- unique(edge_info_all$edge_center_dt$normal_cell_type)\n",
    "combined_fisher_dt <- data.table()\n",
    "for (cell_type_ in cell_types_present) {\n",
    "    temp_dt <- edge_info_all_merged[normal_cell_type == cell_type_,]\n",
    "    ordered_N_dt <- meta_data_dt[cluster == cell_type_,.N,by=sample][order(-N)]\n",
    "    ordered_N_edge_dt <- temp_dt[cell_category==\"edge\",.N,by=sample][,.(sample,N_edge=N)]\n",
    "    ordered_N_dt <- merge( ordered_N_dt, ordered_N_edge_dt, by=\"sample\" )[order(-N),]\n",
    "    \n",
    "    N_edge <- sum(ordered_N_dt$N_edge)\n",
    "    N_total <- sum(ordered_N_dt$N)\n",
    "    \n",
    "    ordered_samples <- ordered_N_dt$sample\n",
    "\n",
    "    p_values <- c()\n",
    "    odds_values <- c()\n",
    "    for (sample_ in ordered_samples) {\n",
    "        N_edge_from_sample <- ordered_N_dt[sample == sample_,N_edge]\n",
    "        N_non_edge_from_sample <- ordered_N_dt[sample == sample_,N] - N_edge_from_sample\n",
    "        N_edge_from_other_samples <- N_edge - N_edge_from_sample\n",
    "        N_non_edge_from_other_samples <- N_total - N_edge_from_sample - N_non_edge_from_sample - N_edge_from_other_samples\n",
    "        odds_mat <- matrix(c(N_edge_from_sample,N_non_edge_from_sample,\n",
    "                             N_edge_from_other_samples,N_non_edge_from_other_samples),byrow=T,nrow=2,ncol=2)\n",
    "        f_test <- fisher.test(odds_mat,alternative=\"greater\")\n",
    "        p_values <- c(p_values,f_test$p.value)\n",
    "        odds_values <- c(odds_values,f_test$estimate)\n",
    "    }\n",
    "    ordered_N_dt[,`:=`(p_value=p_values,normal_cell_type=cell_type_,odds_ratio=odds_values)]\n",
    "    combined_fisher_dt <- rbind( combined_fisher_dt, ordered_N_dt )\n",
    "  \n",
    "}\n",
    "combined_fisher_dt$adj_p_value <- p.adjust(combined_fisher_dt$p_value,method=\"bonferroni\")\n",
    "combined_fisher_dt[adj_p_value < 0.1,]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_info_list <- list()\n",
    "samples <- unique( meta_data_dt$sample )\n",
    "for (sample_id in samples) {\n",
    "    sample_seurat_obj <- subset( seurat_obj, cells=meta_data_dt[sample == sample_id,cell.name])\n",
    "    sample_seurat_obj <- NormalizeData(sample_seurat_obj) %>% FindVariableFeatures(.,nfeatures=1000) %>% \n",
    "    ScaleData(.) %>% RunPCA(.,npcs=50,verbose=F)  %>% RunUMAP(.,dims=1:50,verbose=F)\n",
    "    \n",
    "    homotypic.prop <- modelHomotypic(sample_seurat_obj@meta.data$cluster)\n",
    "    nExp_poi <- round(0.05*nrow(sample_seurat_obj@meta.data))\n",
    "    #nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop)) \n",
    "    \n",
    "    sample_param_sweep <- paramSweep_v3(sample_seurat_obj, PCs = 1:50, sct = FALSE)\n",
    "    sample_param_sweep_summary <- summarizeSweep(sample_param_sweep, GT = FALSE)\n",
    "    sweep_dt <- data.table( find.pK( sample_param_sweep_summary ) )\n",
    "    optimal_pK <- as.double(as.vector(sweep_dt[order(-BCmetric),][1]$pK))\n",
    "    sample_seurat_obj <- doubletFinder_v3(sample_seurat_obj, \n",
    "                               PCs = 1:50, pN = 0.25, pK = optimal_pK, nExp = nExp_poi, \n",
    "                                          reuse.pANN = FALSE, sct = FALSE)\n",
    "    \n",
    "    sample_meta_data_dt <- data.table( sample_seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "    doublet_class_col <- paste(\"DF.classifications_0.25\",optimal_pK,nExp_poi,sep=\"_\")\n",
    "    doublet_info_list[[sample_id]] <- sample_meta_data_dt[,c(\"cell.name\",doublet_class_col),with=F] %>%  setnames(.,doublet_class_col,\"doublet_class\")\n",
    "    #4 + \"\"\n",
    "    #meta_data_dt <- data.table( subset_seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "}\n",
    "\n",
    "merged_doublet_dt <- rbindlist(doublet_info_list)\n",
    "cell_doublets <- meta_data_dt[cell.name %in% merged_doublet_dt[doublet_class == \"Doublet\",cell.name],cell.name]\n",
    "fwrite( data.table(cell.name=cell_doublets), \"cell_doublets.tsv\", sep=\"\\t\", col.names=F, quote=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison with embryonic progenitor genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_exp_dt_list <- list()\n",
    "\n",
    "for (cell_type in c(\"Acinar cell\",\"Ductal cell type 1\")) {\n",
    "    edge_center_dt <- edge_info_all$edge_center_dt[normal_cell_type == cell_type,]\n",
    "\n",
    "    diff_exp_dt_list[[cell_type]] <- get_edge_de_genes_dt( seurat_obj, \n",
    "                                          edge_center_dt                \n",
    "                                        )[p_val_adj < 0.01,]\n",
    "}\n",
    "\n",
    "derived_signatures <- list()\n",
    "\n",
    "all_genes <- rownames(seurat_obj[[\"RNA\"]]@counts)\n",
    "common_edge_genes <- intersect( diff_exp_dt_list[[\"Acinar cell\"]]$gene_name, \n",
    "                               diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name )\n",
    "acinar_only_edge_genes <- setdiff( diff_exp_dt_list[[\"Acinar cell\"]]$gene_name, common_edge_genes )\n",
    "ductal_only_edge_genes <- setdiff( diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name, common_edge_genes )\n",
    "all_edge_genes <- union( diff_exp_dt_list[[\"Acinar cell\"]]$gene_name, \n",
    "                               diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name )\n",
    "\n",
    "derived_signatures[[\"Edge Acinar All\"]] <- diff_exp_dt_list[[\"Acinar cell\"]]$gene_name\n",
    "derived_signatures[[\"Edge Acinar Unique\"]] <- acinar_only_edge_genes\n",
    "derived_signatures[[\"Outlier Ductal All\"]] <- diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name\n",
    "derived_signatures[[\"Outlier Ductal Unique\"]] <- ductal_only_edge_genes\n",
    "derived_signatures[[\"Outlier Universe\"]] <- all_edge_genes\n",
    "derived_signatures[[\"Edge Outlier Common\"]] <- common_edge_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_table <- fread(\"progenitor_marker_table.tsv\")[,!c(\"Stainings\\ and\\ References\")]\n",
    "set(marker_table,NULL,\"Acinar -> Progenitor\",paste(marker_table$`Acinar cell`, marker_table$`Embryonic progenitor`,sep=\" to \"))\n",
    "set(marker_table,NULL,\"Ductal -> Progenitor\",paste(marker_table$`Duct cell`, marker_table$`Embryonic progenitor`,sep=\" to \"))\n",
    "marker_table <- marker_table[,!c(\"Acinar cell\",\"Duct cell\",\"Embryonic progenitor\",\"\")]\n",
    "marker_genes <- marker_table$Marker\n",
    "\n",
    "new_markers <- c(\"STAT3\",\"SEL1L\",\"CBL\",\"KLF4\",\"CTNND1\",\"LKB1\",\"ICAM1\",\"DCLK1\",\"CDKN1A\")\n",
    "marker_genes <- unique( c(marker_genes,new_markers) )\n",
    "marker_genes <- marker_genes[marker_genes %in% names(edge_non_edge_zscores_list[[\"Acinar cell\"]])]\n",
    "\n",
    "for (gene in marker_genes) {\n",
    "    if (!gene %in% marker_table$Marker){\n",
    "        marker_table <- rbind( marker_table, list(gene,\"- to +\",\"- to +\"))\n",
    "    }\n",
    "}\n",
    "rows_to_update <- 1:nrow(marker_table)\n",
    "for (cell_type in c(\"Acinar cell\",\"Ductal cell type 1\")) {\n",
    "    de_status_vec <- rep(\"\",length(marker_genes))#vector(mode=\"character\",length=length(marker_genes))\n",
    "    names(de_status_vec) <- marker_genes\n",
    "    edge_column_values <- rep(\"N/A\",length(marker_genes))\n",
    "    names(edge_column_values) <- marker_genes\n",
    "\n",
    "    de_marker_genes <- marker_genes[marker_genes %in% diff_exp_dt_list[[cell_type]]$gene_name]\n",
    "    de_status_vec[de_marker_genes] <- \"*\"\n",
    "    edge_column_values[marker_genes] <- paste(round(edge_non_edge_zscores_list[[cell_type]][marker_genes],2),de_status_vec)\n",
    "    #edge_column_values[is.na(edge_column_values)] <- \"N/A\"\n",
    "    edge_column_name <- paste(gsub(\"\\ .*\",\"\",cell_type),\"Non-Edge -> Edge\")\n",
    "    \n",
    "    set(marker_table,i=rows_to_update[marker_table$Marker %in% marker_genes],\n",
    "        edge_column_name,edge_column_values)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "progenitor_count_mat <- read_gene_exp_mat(file.path(base_path,\"GSM4194789_TMM_counts_CPM.csv.gz\"))\n",
    "\n",
    "if (!file.exists(file.path(base_path,\"progenitor_annotation.tsv\"))) {\n",
    "    progenitor_seurat_obj <- CreateSeuratObject( progenitor_count_mat )\n",
    "    progenitor_seurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(progenitor_seurat_obj, pattern = \"^MT-\")\n",
    "    dying_cells <- WhichCells( progenitor_seurat_obj, expression='percent.mt > 10' )\n",
    "    \n",
    "    progenitor_seurat_obj <- NormalizeData(progenitor_seurat_obj) \n",
    "    progenitor_seurat_obj <- ScaleData(progenitor_seurat_obj) %>% FindVariableFeatures(.,nfeatures = 1000) %>% RunPCA(.,dims=50) %>% RunUMAP(.,dims=1:50) \n",
    "\n",
    "    homotypic.prop <- modelHomotypic(progenitor_seurat_obj@meta.data$cluster)\n",
    "    nExp_poi <- round(0.05*nrow(progenitor_seurat_obj@meta.data))\n",
    "    #nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop)) \n",
    "\n",
    "    sample_param_sweep <- paramSweep_v3(progenitor_seurat_obj, PCs = 1:50, sct = FALSE)\n",
    "    sample_param_sweep_summary <- summarizeSweep(sample_param_sweep, GT = FALSE)\n",
    "    sweep_dt <- data.table( find.pK( sample_param_sweep_summary ) )\n",
    "    optimal_pK <- as.double(as.vector(sweep_dt[order(-BCmetric),][1]$pK))\n",
    "    progenitor_seurat_obj <- doubletFinder_v3(progenitor_seurat_obj, \n",
    "                               PCs = 1:50, pN = 0.25, pK = optimal_pK, nExp = nExp_poi, \n",
    "                                          reuse.pANN = FALSE, sct = FALSE)\n",
    "\n",
    "    sample_meta_data_dt <- data.table( progenitor_seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "    doublet_class_col <- paste(\"DF.classifications_0.25\",optimal_pK,nExp_poi,sep=\"_\")\n",
    "    doublet_info_dt <- sample_meta_data_dt[,c(\"cell.name\",doublet_class_col),with=F] %>%  setnames(.,doublet_class_col,\"doublet_class\")\n",
    "\n",
    "    cell_doublets <- progenitor_meta_data_dt[cell.name %in% doublet_info_dt[doublet_class == \"Doublet\",cell.name],cell.name]\n",
    "    #fwrite( data.table(cell.name=cell_doublets), \"progenitor_cell_doublets.tsv\", sep=\"\\t\", col.names=F, quote=F)\n",
    "    \n",
    "    cells_to_retain <- setdiff( Cells(progenitor_seurat_obj), c(dying_cells,doublet_cells) )\n",
    "    progenitor_seurat_obj <- subset( progenitor_seurat_obj, cells=cells_to_retain )\n",
    "    progenitor_meta_data_dt <- data.table( progenitor_seurat_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "\n",
    "    progenitor_seurat_obj <- NormalizeData(progenitor_seurat_obj) %>% ScaleData(.) %>% \n",
    "    FindVariableFeatures(.,nfeatures = 1000) %>% RunPCA(.,dims=50) %>% FindNeighbors( ., reduction = \"pca\", k.param=30 ) %>%\n",
    "    FindClusters(.) %>% RunUMAP(.,dims=1:50)\n",
    "    progenitor_meta_data_dt <- data.table( progenitor_seurat_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "    fwrite( progenitor_meta_data_dt, \"progenitor_annotation.tsv\", sep=\"\\t\", row.names=F,quote=F)\n",
    "} else {\n",
    "    progenitor_meta_data_dt <- fread(file.path(base_path,\"progenitor_annotation.tsv\"))\n",
    "    progenitor_seurat_obj <- create_full_seurat_object( progenitor_count_mat, progenitor_meta_data_dt )\n",
    "    progenitor_seurat_obj <- NormalizeData( progenitor_seurat_obj )\n",
    "    #Cluster 7 contains Sox9+,PTF1a+ cells\n",
    "    sox9_ptf1a_cells <- progenitor_meta_data_dt[seurat_clusters == 7,cell.name]\n",
    "}\n",
    "\n",
    "\n",
    "progenitor_seurat_obj <- SetIdent( progenitor_seurat_obj, value=\"seurat_clusters\" )\n",
    "progenitor_markers_df <- FindMarkers( progenitor_seurat_obj, ident.1=\"7\")\n",
    "progenitor_markers_dt <- data.table( progenitor_markers_df, keep.rownames=T ) %>% setnames(.,\"rn\",\"gene_name\")\n",
    "progenitor_genes <- progenitor_markers_dt[p_val < 0.1 & avg_logFC > 0,gene_name]\n",
    "progenitor_genes <- progenitor_genes[progenitor_genes %in% all_genes]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cosine_matrix <- function(X,Y) {\n",
    "    norm_X <- sqrt(colSums(X^2))\n",
    "#     print(Y)\n",
    "    norm_Y <- sqrt(colSums(Y^2))\n",
    "    \n",
    "    dimnames_X <- list(rownames(X),colnames(X))\n",
    "    dimnames_Y <- list(rownames(Y),colnames(Y))\n",
    "\n",
    "    X <- sapply(1:ncol(X),function(idx){return(X[,idx]/norm_X[idx])})\n",
    "    Y <- sapply(1:ncol(Y),function(idx){return(Y[,idx]/norm_Y[idx])})\n",
    "    \n",
    "    dimnames(X) <- dimnames_X\n",
    "    dimnames(Y) <- dimnames_Y\n",
    "    \n",
    "    dist_mat <- t(X) %*% Y\n",
    "    return(dist_mat)\n",
    "}\n",
    "\n",
    "compute_cell_wise_cosine_similarity <- function( query_mat, reference_mat, genes_to_use=NULL, \n",
    "                                                dist_from=\"medoid\",return_sd=F ) {\n",
    "    if (!is.null(genes_to_use)) {\n",
    "        query_mat <- query_mat[genes_to_use,]\n",
    "        reference_mat <- reference_mat[genes_to_use,]\n",
    "    } \n",
    "    \n",
    "    query_ref_dist_mat <- get_cosine_matrix( query_mat, reference_mat )\n",
    "    \n",
    "    if (dist_from == \"medoid\") {\n",
    "        ref_dist_mat <- get_cosine_matrix( reference_mat, reference_mat )\n",
    "        medoid <-which.max(rowSums(ref_dist_mat))\n",
    "        distances <- query_ref_dist_mat[,medoid]\n",
    "        to_return <- list(\"dist\"=distances)\n",
    "    } else if (dist_from == \"all\") {\n",
    "        distances <- Matrix::rowMeans(query_ref_dist_mat)\n",
    "        num_ref_cells <- ncol(reference_mat)\n",
    "        sd_distances <- sqrt( (num_ref_cells/(num_ref_cells-1)) * (rowMeans(query_ref_dist_mat^2) - distances^2))\n",
    "        to_return <- list(\"dist\"=distances,\"sd\"=sd_distances)\n",
    "    }\n",
    "    \n",
    "    return(to_return)\n",
    "}\n",
    "\n",
    "plot_similarity_scatter <- function( dist_list_x, dist_list_y ) {\n",
    "    dt <- data.table( x_mean=dist_list_x$dist,\n",
    "                     y_mean=dist_list_y$dist,\n",
    "                      x_sd=dist_list_x$sd,\n",
    "                     y_sd=dist_list_y$sd )\n",
    "    \n",
    "    p <- ggplot( dt ) + geom_pointrange(aes(x=x_mean,y=y_mean,xmin=x_mean-1.96*x_sd,\n",
    "                                            xmax=x_mean+1.96*x_sd),fatten=1,color=\"gray\",\n",
    "                                        size=0.5) +\n",
    "       geom_pointrange(aes(x=x_mean,y=y_mean,ymin=y_mean-1.96*y_sd,ymax=y_mean+1.96*y_sd),fatten=1,color=\"gray\",\n",
    "                       size=0.5) + \n",
    "    geom_abline(slope=1,intercept=0) + geom_point(aes(x=x_mean,y=y_mean),color=\"black\",size=1) +\n",
    "    xlim(0,1) + ylim(0,1)\n",
    " \n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "progenitor_mat <- progenitor_seurat_obj[[\"RNA\"]]@data[progenitor_genes,sox9_ptf1a_cells]\n",
    "\n",
    "acinar_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category==\"edge\",cell.name]\n",
    "acinar_non_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category!=\"edge\",cell.name]\n",
    "\n",
    "ductal_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                     cell_category == \"edge\",cell.name]\n",
    "ductal_non_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                     cell_category != \"edge\",cell.name]\n",
    "\n",
    "malignant_ductal_cells <- meta_data_dt[cluster == \"Ductal cell type 2\",cell.name]\n",
    "\n",
    "# acinar_seurat_obj <- \n",
    "ductal_outlier_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,ductal_outlier_cells]\n",
    "acinar_edge_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,acinar_edge_cells]\n",
    "ductal_non_outlier_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,ductal_non_outlier_cells]\n",
    "acinar_non_edge_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,acinar_non_edge_cells]\n",
    "malignant_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,malignant_ductal_cells]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ret_edge_acinar <- compute_cell_wise_cosine_similarity( acinar_edge_data_mat, progenitor_mat,\n",
    "                                           dist_from=\"all\")\n",
    "ret_outlier_ductal <- compute_cell_wise_cosine_similarity( ductal_outlier_data_mat, progenitor_mat,\n",
    "                                           dist_from=\"all\")\n",
    "\n",
    "ret_non_edge_acinar <- compute_cell_wise_cosine_similarity( acinar_non_edge_data_mat,progenitor_mat,\n",
    "                                           dist_from=\"all\")\n",
    "ret_non_outlier_ductal <- compute_cell_wise_cosine_similarity( ductal_non_outlier_data_mat, progenitor_mat,\n",
    "                                            dist_from=\"all\")\n",
    "\n",
    "\n",
    "edge_acinar_dist_dt <- data.table(value=ret_edge_acinar$dist,variable=\"Edge Acinar\")\n",
    "outlier_ductal_dist_dt <- data.table(value=ret_outlier_ductal$dist,variable=\"Outlier Ductal\")\n",
    "non_edge_acinar_dist_dt <- data.table(value=ret_non_edge_acinar$dist,variable=\"Non-Edge Acinar\")\n",
    "non_outlier_ductal_dist_dt <- data.table(value=ret_non_outlier_ductal$dist,variable=\"Non-Outlier Ductal\")\n",
    "\n",
    "options(repr.plot.width=4,repr.plot.height=6)\n",
    "plot_dt <- rbind(edge_acinar_dist_dt,non_edge_acinar_dist_dt,outlier_ductal_dist_dt,\n",
    "                 non_outlier_ductal_dist_dt)\n",
    "\n",
    "plot_dt$cell_type <- plot_dt$variable\n",
    "plot_dt[grepl(\"Acinar\",variable),]$cell_type <- \"Acinar\"\n",
    "plot_dt[grepl(\"Ductal cell type 1\",variable),]$cell_type <- \"Ductal\"\n",
    "# plot_dt[grepl(\"Ductal cell type 2\",variable),]$cell_type <- \"Ductal\"\n",
    "\n",
    "ggboxplot(plot_dt,x=\"variable\",y=\"value\",color=\"cell_type\") +\n",
    "theme_classic(base_size=15) + ylab(\"Cosine similarity\") + xlab(\"\") + \n",
    "# theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) +\n",
    "stat_compare_means(comparisons=list(c(\"Edge Acinar\",\"Outlier Ductal\"),\n",
    "                                    c(\"Edge Acinar\",\"Non-Edge Acinar\"),\n",
    "                                   c(\"Outlier Ductal\",\"Non-Outlier Ductal\")),step.increase=0.2,label.x=1.5,label=\"p.signif\",size=8) +\n",
    "scale_x_discrete(labels=c(\"Edge Acinar\"=\"E\",\"Outlier Ductal\"=\"O\",\"Non-Edge Acinar\"=\"NE\",\"Non-Outlier Ductal\"=\"NO\")) + \n",
    "labs(color=\"Cell Type\") + theme(legend.position=\"bottom\") + ylim(0,1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6,repr.plot.height=8)\n",
    "plot_dt <- rbind(edge_acinar_dist_dt,non_edge_acinar_dist_dt,outlier_ductal_dist_dt,\n",
    "                 non_outlier_ductal_dist_dt,malignant_dist_dt)\n",
    "\n",
    "plot_dt$cell_type <- plot_dt$variable\n",
    "plot_dt[grepl(\"Acinar\",variable),]$cell_type <- \"Acinar\"\n",
    "plot_dt[grepl(\"Ductal cell type 1\",variable),]$cell_type <- \"Ductal\"\n",
    "plot_dt[grepl(\"Ductal cell type 2\",variable),]$cell_type <- \"Malignant\"\n",
    "\n",
    "ggboxplot(plot_dt,x=\"variable\",y=\"value\",color=\"cell_type\") +\n",
    "theme_classic(base_size=15) + ylab(\"Cosine similarity\") + xlab(\"\") + \n",
    "# theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) +\n",
    "stat_compare_means(comparisons=list(c(\"Edge Acinar\",\"Outlier Ductal\"),\n",
    "                                    c(\"Edge Acinar\",\"Non-Edge Acinar\"),\n",
    "                                   c(\"Outlier Ductal\",\"Non-Outlier Ductal\"),\n",
    "                                   c(\"Malignant Ductal\",\"Edge Acinar\")),label.x=1.5,label=\"p.signif\",size=8) +\n",
    "scale_x_discrete(labels=c(\"Edge Acinar\"=\"E\",\"Outlier Ductal\"=\"O\",\"Non-Edge Acinar\"=\"NE\",\n",
    "                          \"Non-Outlier Ductal\"=\"NO\",\"Malignant Ductal\"=\"Mal\")) + \n",
    "labs(color=\"Cell Type\") + theme(legend.position=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FGSEA enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_h <- gmtPathways(file.path(base_path,\"h.all.v7.0.symbols.gmt\"))\n",
    "\n",
    "all_genes <- rownames(seurat_obj[[\"RNA\"]]@data)\n",
    "\n",
    "cancersea_gene_sets <- list()\n",
    "signature_file_paths <- list.files(file.path(base_path,\"cancer-sea\"),full.names=T)\n",
    "for (idx in 1:length(signature_file_paths)) {\n",
    "    signature_genes_dt <- fread(signature_file_paths[idx])\n",
    "    temp <- unlist(strsplit(signature_file_paths[idx],\"\\\\.\"))[1] \n",
    "    signature_name <- paste( \"CancerSEA\", tail( unlist( strsplit( temp, \"/\" ) ), 1 ), sep=\"_\" )\n",
    "    signature_gene_names <- signature_genes_dt$GeneName\n",
    "    cancersea_gene_sets[[signature_name]] <- signature_gene_names\n",
    "}\n",
    "pathways_all <- c(pathways_h,cancersea_gene_sets)\n",
    "\n",
    "num_pathways <- length(pathways_all)\n",
    "\n",
    "pathways_all <- lapply( pathways_all, function(pathway_genes) {return(pathway_genes[pathway_genes %in% all_genes])})\n",
    "for (pathway in names(pathways_all)) {\n",
    "    if (length(pathways_all[[pathway]]) == 0) {\n",
    "        pathways_all[[pathway]] <- NULL\n",
    "    }\n",
    "} \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conventional GSEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normal_cell_types <- c(\"Acinar cell\",\"Ductal cell type 1\")\n",
    "malignant_cell_type <- \"Ductal cell type 2\"\n",
    "malignant_cells <- meta_data_dt[cluster == malignant_cell_type,cell.name]\n",
    "fgsea_dt <- data.table(pathway=names(pathways_all))\n",
    "\n",
    "comparison_list <- list(\"Malignant v All Acinar\"=c(\"Ductal cell type 2\",\"Acinar cell\"),\n",
    "                       \"Edge v Non-Edge Acinar\"=c(\"edge Acinar cell\",\"center Acinar cell\"),\n",
    "                        \"Malignant v All Ductal\"=c(\"Ductal cell type 2\",\"Ductal cell type 1\"),\n",
    "                       \"Outlier v Non-Outlier Ductal\"=c(\"edge Ductal cell type 1\",\"center Ductal cell type 1\"))\n",
    "\n",
    "get_cells <- function(cluster_name) {\n",
    "    if (!grepl(\"edge\",cluster_name) && !grepl(\"center\",cluster_name)) {\n",
    "        cells <- meta_data_dt[cluster == cluster_name,cell.name]\n",
    "    } else {\n",
    "        if (grepl(\"edge\",cluster_name)) {\n",
    "            cell_category_ <- \"edge\"\n",
    "        } else {\n",
    "            cell_category_ <- \"center\"\n",
    "        }\n",
    "        cluster_name <- gsub(paste(cell_category_,\"\"),\"\",cluster_name)\n",
    "        cells <- edge_info_all$edge_center_dt[normal_cell_type == cluster_name & cell_category == cell_category_,\n",
    "                                             cell.name]\n",
    "    }\n",
    "    \n",
    "    return(cells)\n",
    "}\n",
    "\n",
    "\n",
    "for (comparison in names(comparison_list)) {\n",
    "    print(comparison)\n",
    "    flush.console()\n",
    "    column_pair <- comparison_list[[comparison]]\n",
    "    cells_1 <- get_cells( column_pair[1] )\n",
    "    cells_2 <- get_cells( column_pair[2] )\n",
    "    cells <- union(cells_1,cells_2)\n",
    "    subset_seurat_obj <- subset( seurat_obj, cells=cells ) %>% ScaleData(.)\n",
    "    scaled_data <- subset_seurat_obj[[\"RNA\"]]@scale.data\n",
    "    \n",
    "    zscore_diff <- rowMeans(scaled_data[,cells_1]) - rowMeans(scaled_data[,cells_2])\n",
    "\n",
    "    dt <- run_fgsea( pathways_all, zscore_diff)\n",
    "    fgsea_dt <- merge( fgsea_dt, dt[,.(pathway,padj,NES)],by=\"pathway\") %>% \n",
    "    setnames(.,\"padj\",paste(\"padj\",comparison)) %>% setnames(.,\"NES\",paste(\"NES\",comparison))\n",
    "    rm(scaled_data)\n",
    "    rm(subset_seurat_obj)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "make_pathway_annotation <- function(disp_mat,num_clusters) {\n",
    "     pathway_hclust_obj <- hclust( dist(t(scale(t(disp_mat)))) )\n",
    "#     pathway_hclust_obj <- hclust( dist(disp_mat), method=\"complete\" )\n",
    "\n",
    "    pathway_clusters <- cutree(tree = pathway_hclust_obj, k = num_clusters)\n",
    "    cluster_colors <- list()\n",
    "    cluster_colors[[\"Cluster\"]] <- brewer.pal(n = num_clusters, name = \"Accent\")\n",
    "    names(cluster_colors[[\"Cluster\"]]) <- paste(\"Cluster\", 1:num_clusters)\n",
    "\n",
    "    pathway_cluster_dt <- data.table( `Gene Set`=paste( \"Cluster\", pathway_clusters ) )\n",
    "    rownames(pathway_cluster_dt) <- names(pathway_clusters)\n",
    "    \n",
    "    return(pathway_cluster_dt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_sig <- function(col) {\n",
    "    row_mask <- fgsea_dt[[paste(\"padj\", col)]] < 0.05\n",
    "    #idxes <- 1:nrow(fgsea_dt)\n",
    "    return(row_mask)\n",
    "}\n",
    "cluster_dt <- data.table(pathway=fgsea_dt$pathway,`Enriched in`=\"C8:Misc.\")\n",
    "\n",
    "cluster_dt[get_sig(\"Edge v Non-Edge Acinar\"),`Enriched in`:=\"C6:EA-NEA\"]\n",
    "cluster_dt[get_sig(\"Outlier v Non-Outlier Ductal\"),`Enriched in`:=\"C7:OD-NOD\"]\n",
    "cluster_dt[get_sig(\"Malignant v All Acinar\"),`Enriched in`:=\"C3:Mal-Aci\"]\n",
    "cluster_dt[get_sig(\"Malignant v All Ductal\"),`Enriched in`:=\"C4:Mal-Duc\"]\n",
    "\n",
    "\n",
    "cluster_dt[get_sig(\"Malignant v All Ductal\") & get_sig(\"Malignant v All Acinar\"),\n",
    "                   `Enriched in`:=\"C2:Mal-Duc & Mal-Aci\"]\n",
    "cluster_dt[get_sig(\"Edge v Non-Edge Acinar\") & get_sig(\"Outlier v Non-Outlier Ductal\"),\n",
    "           `Enriched in`:=\"C5:EA-NEA & OD-NOD\"]\n",
    "cluster_dt[get_sig(\"Malignant v All Ductal\") & get_sig(\"Malignant v All Acinar\") &\n",
    "                get_sig(\"Edge v Non-Edge Acinar\") & get_sig(\"Outlier v Non-Outlier Ductal\"),\n",
    "                   `Enriched in`:=\"C1:All\"]\n",
    "\n",
    "cluster_dt <- cluster_dt[order(`Enriched in`),]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row_order <- c(\"C1:All\",\"C2:Mal-Duc & Mal-Aci\",\n",
    "                                        \"C3:Mal-Aci\",\n",
    "                                        \"C4:Mal-Duc\",\n",
    "                                       \"C5:EA-NEA & OD-NOD\",\n",
    "                                       \"C6:EA-NEA\",\n",
    "                                       \"C7:OD-NOD\")\n",
    "                                         #\"C8:Misc.\")\n",
    "pathway_cluster_dt <- data.table()\n",
    "gap_rows <- c()\n",
    "idx <- 0\n",
    "new_row_entries <- c()\n",
    "for (row in row_order) {\n",
    "    if (row == \"Misc.\")\n",
    "        next\n",
    "    dt <- cluster_dt[`Enriched in` == row,]\n",
    "    dt$`Enriched in` <- paste(row, paste0(\"(\",nrow(dt),\")\"))\n",
    "    pathway_cluster_dt <- rbind( pathway_cluster_dt, dt)\n",
    "    idx <- idx + nrow(dt)\n",
    "    gap_rows <- c(gap_rows,idx)\n",
    "    new_row_entries <- c(new_row_entries,dt$`Enriched in`)\n",
    "}\n",
    "pathway_cluster_dt$`Enriched in` <- factor(pathway_cluster_dt$`Enriched in`,levels=unique(new_row_entries))\n",
    "pathway_order <- pathway_cluster_dt$pathway\n",
    "pathway_cluster_dt <- pathway_cluster_dt[,!c(\"pathway\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "theme_set(theme_gray(base_size = 8))\n",
    "\n",
    "all_col_names <- names(fgsea_dt)\n",
    "nes_cols <- all_col_names[grepl(\"NES\",all_col_names)]\n",
    "padj_cols <- all_col_names[grepl(\"padj\",all_col_names)]\n",
    "\n",
    "# sig_pathways <- melt(fgsea_dt[,c(\"pathway\",padj_cols),with=F],id.vars=\"pathway\")[value < 0.05,pathway] %>% unique(.)\n",
    "fgsea_dt <- fgsea_dt[order(match(pathway,pathway_order)),]\n",
    "pathway_mat <- as.matrix(fgsea_dt[,nes_cols,with=F])\n",
    "colnames(pathway_mat) <- gsub(\"NES \",\"\", nes_cols )\n",
    "rownames(pathway_mat) <- fgsea_dt$pathway\n",
    "\n",
    "flip_to_NA <- function(x) {\n",
    "    nes_scores <- pathway_mat[,x]\n",
    "    p_vals <- fgsea_dt[[paste(\"padj\",x)]]\n",
    "    nes_scores[p_vals > 0.05] = NA\n",
    "#      nes_scores[nes_scores > 0] <- 1\n",
    "#      nes_scores[nes_scores < 0] <- -1\n",
    "     return(nes_scores)\n",
    "}\n",
    "\n",
    "na_pathway_mat <- sapply( names(comparison_list), flip_to_NA )\n",
    "colnames(na_pathway_mat) <- colnames(pathway_mat)\n",
    "rownames(pathway_cluster_dt) <- pathway_order\n",
    "na_pathway_mat <- na_pathway_mat[pathway_order,]\n",
    "p <- pheatmap(na_pathway_mat,fontsize=15,cluster_cols=F,cluster_rows=F,show_rownames=T,\n",
    "         annotation_row=pathway_cluster_dt, na_col=\"gray\",border_color=\"black\",\n",
    "              colorRampPalette(rev(brewer.pal(n = 8, name = \"RdYlBu\")))(4),\n",
    "             breaks=c(-2,-1,0,1,2),\n",
    "              gaps_row=gap_rows,cellheight=15,cellwidth=15,\n",
    "             filename=\"enrichment_heatmap.png\",width=20,height=20,dpi=300)\n",
    "             "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tumour-wise scoring of ductal and acinar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acinar_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category==\"edge\",cell.name]\n",
    "acinar_non_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category!=\"edge\",cell.name]\n",
    "\n",
    "ductal_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                     cell_category != \"edge\",cell.name]\n",
    "ductal_non_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                      cell_category != \"edge\",cell.name]\n",
    "\n",
    "ductal_non_outlier_data_mat <- seurat_obj[[\"RNA\"]]@data[,ductal_non_outlier_cells]\n",
    "acinar_non_edge_data_mat <- seurat_obj[[\"RNA\"]]@data[,acinar_non_edge_cells]\n",
    "\n",
    "ret_edge_acinar <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, acinar_edge_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Edge Acinar All\"]], dist_from=\"medoid\")\n",
    "ret_outlier_ductal <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, ductal_outlier_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Outlier Ductal All\"]], dist_from=\"medoid\")\n",
    "ret_non_edge_acinar <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, acinar_non_edge_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Edge Acinar All\"]], dist_from=\"medoid\")\n",
    "ret_non_outlier_ductal <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, ductal_non_outlier_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Outlier Ductal All\"]], dist_from=\"medoid\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=16,repr.plot.height=6)\n",
    "melted_dt <- melt(data.table(`Edge Acinar`=ret_edge_acinar$dist,\n",
    "                        `Outlier Ductal`=ret_non_outlier_ductal$dist,\n",
    "                        `Non-Edge Acinar`=ret_non_edge_acinar$dist,\n",
    "                        `Non-Outlier Ductal`=ret_non_outlier_ductal$dist,\n",
    "                       cell.name=names(ret_edge_acinar$dist)),id.vars=c(\"cell.name\")\n",
    "                 ) #%>% setnames(.,\"variable\",\"Cell Type\")\n",
    "# 4 + \"\"\n",
    "melted_dt <- merge( melted_dt, meta_data_dt[,.(sample,cell.name)])\n",
    "melted_dt$sample <- factor(melted_dt$sample,levels=paste0(\"T\",1:24))\n",
    "\n",
    "p_edge_outlier <- ggplot( melted_dt[!grepl(\"Non-\",variable),]  ) + \n",
    "geom_boxplot(aes(y=value,x=sample,color=variable),size=1.0,fill=NA,width=0.5) + \n",
    "# stat_compare_means(aes(x=variable,y=value),label.x=1.5,label=\"p.signif\",size=8) + \n",
    "xlab(\"\") + ylab(\"\") +\n",
    "theme_classic(base_size=20) + labs(color=\"\") + theme(legend.position=c(0.5,0.1), legend.direction=\"horizontal\")\n",
    "\n",
    "x_labels <- 1:24\n",
    "names(x_labels) <- paste0(\"T\",x_labels)\n",
    "p_all <- ggplot( melted_dt  ) + \n",
    "geom_boxplot(aes(y=value,x=sample,color=variable),size=1.0,fill=NA,width=0.5) + \n",
    "theme_classic(base_size=20) + ylab(\"Cosine similarity\") + xlab(\"\") + labs(color=\"Cell Type\") +\n",
    "ylim(0.3,1) + theme(legend.direction=\"horizontal\",legend.position=c(0.5,0.1))\n",
    "\n",
    "\n",
    "melted_dt$cell_type <- melted_dt$variable\n",
    "melted_dt[grepl(\"Acinar\",variable),]$cell_type <- \"Acinar\"\n",
    "melted_dt[grepl(\"Ductal\",variable),]$cell_type <- \"Ductal\"\n",
    "# ggplot( melted_dt ) + \n",
    "# geom_boxplot(aes(y=value,group=variable),size=1.0,fill=NA,width=0.5) + \n",
    "p_pan_tumor <- ggboxplot(melted_dt,x=\"variable\",y=\"value\",color=\"cell_type\") +\n",
    "theme_classic(base_size=20) + ylab(\"Cosine similarity\") + xlab(\"\") + \n",
    "# theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) +\n",
    "stat_compare_means(comparisons=list(c(\"Edge Acinar\",\"Outlier Ductal\"),\n",
    "                                    c(\"Edge Acinar\",\"Non-Edge Acinar\"),\n",
    "                                   c(\"Outlier Ductal\",\"Non-Outlier Ductal\")),\n",
    "                   label.x=1.5,label=\"p.signif\",size=5) +\n",
    "scale_x_discrete(labels=c(\"Edge Acinar\"=\"E\",\"Outlier Ductal\"=\"O\",\"Non-Edge Acinar\"=\"NE\",\"Non-Outlier Ductal\"=\"NO\")) + \n",
    "scale_fill_discrete(name = c(\"cell_type\"=\"Cell Type\")) + labs(color=\"\") + \n",
    "theme(legend.position=\"none\") #+ ylim(0.3,0.9)\n",
    "\n",
    "print(\"# of tumors where edge acinar is closer to malignant than outlier ductal\")\n",
    "p_values <- c()\n",
    "for (sample_num in unique(melted_dt$sample)) {\n",
    "    dt <- melted_dt[sample == sample_num,]\n",
    "    p_value <- wilcox.test(dt[variable == \"Edge Acinar\",value],dt[variable == \"Outlier Ductal\",value])$p.value\n",
    "    p_values <- c(p_values,p_value)\n",
    "}\n",
    "adj_p <- p.adjust(p_values,method=\"bonferroni\")\n",
    "print(sum(adj_p < 0.1))\n",
    "\n",
    "rm(melted_dt)\n",
    "rm(ductal_outlier_data_mat)\n",
    "rm(ductal_non_outlier_data_mat)\n",
    "rm(acinar_edge_data_mat)\n",
    "rm(acinar_non_edge_data_mat)\n",
    "\n",
    "final <- ggarrange(p_pan_tumor,p_edge_outlier,nrow=1,ncol=2,widths=c(0.3,1),labels=\"auto\")\n",
    "print(final)\n",
    "print(p_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mechanisms of edge-ness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"../src/motif_analysis.R\")\n",
    "atac_footprints_dt <- fread(file.path(base_path,\"match\",\"footprints_mpbs.bed\"))[,!c(\"V6\",\"V7\")] %>%\n",
    "setnames(.,paste0(\"V\",1:5),c(\"chr\",\"start\",\"end\",\"motif\",\"motif_score\")) %>% unique(.)\n",
    "\n",
    "atac_peaks_dt <- fread(file.path(base_path,\"acinar_atac_peaks.narrowPeak\"))[,.(chr=V1,start=V2,end=V3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                           cell_category == \"edge\",cell.name]\n",
    "non_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                               cell_category != \"edge\",cell.name]\n",
    "\n",
    "acinar_obj <- subset( seurat_obj, cells=c(edge_cells,non_edge_cells) )\n",
    "acinar_obj <- ScaleData(acinar_obj)\n",
    "\n",
    "\n",
    "all_genes <- rownames(seurat_obj[[\"RNA\"]]@data)\n",
    "entrez_dt <- get_entrez_dt( all_genes )\n",
    "entrez_ids <- entrez_dt$final_entrez_id\n",
    "acinar_edge_up_genes <- diff_exp_dt_list[[\"Acinar cell\"]][avg_logFC > 0,gene_name]\n",
    "entrez_dt[SYMBOL %in% acinar_edge_up_genes,`:=`(gene_type=\"Edge-Up\")]\n",
    "entrez_dt[is.na(gene_type),gene_type:=\"Remaining\"]\n",
    "cds_tx_dt <- select(TxDb.Hsapiens.UCSC.hg19.knownGene,columns=c(\"GENEID\",\"TXNAME\"),\n",
    "               keytype=c(\"GENEID\"), keys=entrez_ids ) %>% na.omit(.)\n",
    "cds_tx_dt <- merge( cds_tx_dt, entrez_dt[,.(SYMBOL,final_entrez_id,gene_type)],\n",
    "                   by.x=\"GENEID\", by.y=\"final_entrez_id\") %>% data.table(.)\n",
    "\n",
    "\n",
    "gene_assign <- function( regulatory_dt, assign_type=\"overlap\", dist_threshold=8000 ) {\n",
    "    granges_obj <- makeGRangesFromDataFrame( regulatory_dt, keep.extra.columns=T)\n",
    "\n",
    "    promoters_txdb <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene,upstream=dist_threshold,downstream=400) %>% trim(.)\n",
    "\n",
    "    mcol_df <- as.data.frame(mcols(promoters_txdb))\n",
    "    mcol_df$gene_type <- \"Remaining\"\n",
    "    mcol_df[mcol_df$tx_name %in% cds_tx_dt[gene_type == \"Edge-Up\",TXNAME],]$gene_type <- \"Edge-Up\"\n",
    "    mcols(promoters_txdb) <- mcol_df\n",
    "    rm(mcol_df) \n",
    "\n",
    "    promoter_dt <- data.table( as.data.frame(promoters_txdb))\n",
    "    \n",
    "    overlap_obj <- findOverlaps( granges_obj, promoters_txdb )\n",
    "    regulatory_dt[queryHits(overlap_obj),`:=`(TXNAME=promoter_dt[subjectHits(overlap_obj),tx_name])]\n",
    "\n",
    "    print(head(regulatory_dt))\n",
    "    print(head(cds_tx_dt))\n",
    "    regulatory_dt <- merge( regulatory_dt, cds_tx_dt[,.(TXNAME,gene_type,SYMBOL)], by=\"TXNAME\", all.x=T )[,!c(\"TXNAME\")]\n",
    "    return(regulatory_dt)\n",
    "}\n",
    "atac_peaks_dt <- gene_assign( atac_peaks_dt, assign_type=\"overlap\", dist_threshold = 10000 )\n",
    "# atac_footprints_dt <- gene_assign( atac_footprints_dt, assign_type=\"overlap\",dist_threshold=10000 )\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_atac_peaks_dt <- atac_peaks_dt[!is.na(SYMBOL),c(\"chr\",\"start\",\"end\",\"SYMBOL\",\"gene_type\")]#gene_assign( atac_peaks_dt, assign_type=\"overlap\", dist_threshold = 10000 )\n",
    "peaks_not_near_promoter_dt <- atac_peaks_dt[is.na(SYMBOL),c(\"chr\",\"start\",\"end\",\"SYMBOL\",\"gene_type\")]\n",
    "\n",
    "promoter_atac_granges <- makeGRangesFromDataFrame( promoter_atac_peaks_dt ) %>% resize(.,width=100,fix=\"center\")\n",
    "atac_not_near_promoter_granges <- makeGRangesFromDataFrame( peaks_not_near_promoter_dt ) %>% resize(.,width=100,fix=\"center\")\n",
    "#\n",
    "\n",
    "edge_with_non_edge_list <- pwm_scan( merged_pwm_list, \n",
    "                             foreground_granges_obj = promoter_atac_granges, \n",
    "                             background_granges_obj = atac_not_near_promoter_granges,\n",
    "                             motif_p_value_thresh = 1e-4 )\n",
    "\n",
    "enrichment_dt <- edge_with_non_edge_list$enrichment_dt\n",
    "tfs <- unique(enrichment_dt$tf)\n",
    "foreground_dt <- edge_with_non_edge_list$fg_matches_dt\n",
    "foreground_dt[,`:=`(gene_type=promoter_atac_peaks_dt$gene_type,SYMBOL=promoter_atac_peaks_dt$SYMBOL)]\n",
    "\n",
    "gene_exp_mat <- acinar_obj[[\"RNA\"]]@data[unique(promoter_atac_peaks_dt$SYMBOL),]\n",
    "edge_non_edge_diff_exp <- rowMeans(gene_exp_mat[,edge_cells]) - rowMeans(gene_exp_mat[,non_edge_cells])\n",
    "\n",
    "bound_edge_genes <- promoter_atac_peaks_dt[gene_type == \"Edge\",SYMBOL] %>% unique(.)\n",
    "bound_non_edge_genes <- promoter_atac_peaks_dt[gene_type != \"Edge\",SYMBOL] %>% unique(.)\n",
    "bound_genes <- c(bound_edge_genes,bound_non_edge_genes)\n",
    "\n",
    "num_replicates <- 1000\n",
    "logFC_random_vec <- vector(mode=\"numeric\",length=num_replicates)\n",
    "\n",
    "idx <- 1\n",
    "\n",
    "for (tf_ in tfs) {\n",
    "    print(idx)\n",
    "    flush.console()\n",
    "    tf_dt <- foreground_dt[get(tf_) == 1,c(tf_,\"gene_type\",\"SYMBOL\"),with=F] %>% unique(.)\n",
    "\n",
    "    nearby_edge_genes <- unique(tf_dt[gene_type == \"Edge\",SYMBOL])\n",
    "    num_motif_edge <- length( nearby_edge_genes )\n",
    "    nearby_non_edge_genes <- unique(tf_dt[gene_type != \"Edge\",SYMBOL])\n",
    "\n",
    "    num_motif_non_edge <- length( nearby_non_edge_genes )\n",
    "    motif_bound_genes <- c(nearby_edge_genes,nearby_non_edge_genes)    \n",
    "\n",
    "    edge_logFC <- mean(edge_non_edge_diff_exp[motif_bound_genes])\n",
    "    logFC_p_value_ <- 1\n",
    "    if (num_motif_edge > 0 & num_motif_non_edge > 0) {\n",
    "        for (rep in 1:num_replicates) {\n",
    "            random_edge_genes <- sample( bound_edge_genes, size=num_motif_edge) \n",
    "            random_non_edge_genes <- sample( bound_non_edge_genes, size=num_motif_non_edge )\n",
    "            logFC_random_vec[rep] <- mean(edge_non_edge_diff_exp[c(random_edge_genes,random_non_edge_genes)])\n",
    "        }\n",
    "        \n",
    "        m <- mean(logFC_random_vec)\n",
    "        s <- sd(logFC_random_vec)\n",
    "        y <- dnorm(logFC_random_vec,mean=m,sd=s)\n",
    "        logFC_p_value_ <- 1 - pnorm(edge_logFC,mean=m,sd=s)\n",
    "    } \n",
    "    \n",
    "    enrichment_dt[tf == tf_,`:=`(logFC=edge_logFC,logFC_p_value=logFC_p_value_)]\n",
    "    idx <- idx+1\n",
    "\n",
    "}\n",
    "enrichment_dt[,logFC_p_adj:=p.adjust(logFC_p_value) ]\n",
    "sig_dt <- enrichment_dt[logFC_p_adj < 0.1 & q_value < 0.1,][order(-fold_enrichment),.(tf,fold_enrichment,logFC)]\n",
    "fwrite(sig_dt,\"tfs_enriched_logFC.tsv\",sep=\"\\t\",quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs <- unique(atac_footprints_dt$motif)\n",
    "\n",
    "# overlap_obj <- findOverlaps( makeGRangesFromDataFrame(atac_footprints_dt), \n",
    "#                            makeGRangesFromDataFrame(atac_peaks_dt) )\n",
    "# atac_footprints_dt <- cbind( atac_footprints_dt[queryHits(overlap_obj),],\n",
    "#                   atac_peaks_dt[subjectHits(overlap_obj),.(nearest_gene=SYMBOL)] )\n",
    "\n",
    "promoter_atac_footprints_dt <- peaks_with_footprints_dt[!is.na(nearest_gene),\n",
    "                                                  c(\"chr\",\"start\",\"end\",\n",
    "                                                    \"motif\",\"gene_type\",\"nearest_gene\")]\n",
    "footprints_not_near_promoter_dt <- peaks_with_footprints_dt[is.na(nearest_gene),c(\"chr\",\"start\",\"end\",\"motif\",\"gene_type\",\"nearest_gene\")]\n",
    "p_value_dt <- data.table( motif=motifs, motif_p_value=1 )\n",
    "frac_occ_dt <- data.table( motif=motifs )\n",
    "num_edge_genes <- length(diff_exp_dt_list[[\"Acinar cell\"]][avg_logFC > 0,gene_name])#length(unique(atac_footprints_dt[gene_type == \"Edge\",nearest_gene]))\n",
    "num_non_edge_genes <- length(all_genes) - num_edge_genes\n",
    "edge_gene_sets <- list()\n",
    "non_edge_gene_sets <- list()\n",
    "gene_exp_mat <- acinar_obj[[\"RNA\"]]@data[unique(atac_footprints_dt[!is.na(nearest_gene),nearest_gene]),]\n",
    "edge_non_edge_diff_exp <- rowMeans(gene_exp_mat[,edge_cells]) - rowMeans(gene_exp_mat[,non_edge_cells])\n",
    "\n",
    "bound_edge_genes <- atac_footprints_dt[gene_type == \"Edge-Up\",nearest_gene] %>% unique(.)\n",
    "bound_non_edge_genes <- atac_footprints_dt[gene_type == \"Remaining\",nearest_gene] %>% unique(.)\n",
    "bound_genes <- c(bound_edge_genes,bound_non_edge_genes)\n",
    "\n",
    "num_replicates <- 1000\n",
    "logFC_random_vec <- vector(mode=\"numeric\",length=num_replicates)\n",
    "\n",
    "idx <- 1\n",
    "for (motif_ in motifs) {\n",
    "    print(paste(idx,length(motifs),sep=\"/\"))\n",
    "    flush.console()\n",
    "    promoter_dt <- promoter_atac_footprints_dt[motif == motif_ & !is.na(nearest_gene),]\n",
    "    not_promoter_dt <- footprints_not_near_promoter_dt[motif == motif_,] %>% unique(.)\n",
    "\n",
    "    num_footprints_near_promoter <- nrow( unique(promoter_dt[,.(chr,start,end)]) )\n",
    "    nearby_edge_genes <- unique(promoter_dt[gene_type == \"Edge-Up\",nearest_gene])\n",
    "    nearby_non_edge_genes <- unique(promoter_dt[gene_type == \"Remaining\",nearest_gene])\n",
    "    num_footprints_not_near_promoter <- nrow( unique(not_promoter_dt[,.(chr,start,end)]))\n",
    "    \n",
    "    count_dt <- unique(peaks_with_footprints_dt[motif!=motif_,.(chr,start,end,gene_type)])[,.N,by=gene_type]\n",
    "  \n",
    "    num_no_footprints_near_promoter <- sum(count_dt[!is.na(gene_type),N])\n",
    "    num_no_footprints_not_near_promoter <- count_dt[is.na(gene_type),N]\n",
    "    fisher_mat <- matrix(c(num_footprints_near_promoter,num_footprints_not_near_promoter,\n",
    "                          num_no_footprints_near_promoter,num_no_footprints_not_near_promoter),\n",
    "                        nrow=2,ncol=2,byrow=T)\n",
    "    motif_fisher_p_value <- fisher.test( fisher_mat, alternative = \"g\" )$p.value\n",
    "\n",
    "    p_value_dt[motif==motif_,`:=`(motif_p_value=motif_fisher_p_value)]#,p_value_chi_sq=chisq_p_value)]\n",
    "    motif_bound_genes <- c(nearby_edge_genes,nearby_non_edge_genes)\n",
    "    edge_logFC <- mean(edge_non_edge_diff_exp[motif_bound_genes])\n",
    "    \n",
    "    logFC_p_value_ <- 1\n",
    "    if (num_motif_edge > 0 & num_motif_non_edge > 0) {\n",
    "        for (rep in 1:num_replicates) {\n",
    "            random_edge_genes <- sample( bound_edge_genes, size=num_motif_edge) \n",
    "            random_non_edge_genes <- sample( bound_non_edge_genes, size=num_motif_non_edge )\n",
    "            logFC_random_vec[rep] <- mean(edge_non_edge_diff_exp[c(random_edge_genes,random_non_edge_genes)])\n",
    "        }\n",
    "        \n",
    "        m <- mean(logFC_random_vec)\n",
    "        s <- sd(logFC_random_vec)\n",
    "        y <- dnorm(logFC_random_vec,mean=m,sd=s)\n",
    "        logFC_p_value_ <- 1 - pnorm(edge_logFC,mean=m,sd=s)\n",
    "    } \n",
    "    \n",
    "    frac_occ_dt[motif==motif_,`:=`(promoter_frac=num_footprints_near_promoter/num_no_footprints_near_promoter,\n",
    "                                   non_promoter_frac=num_footprints_not_near_promoter/num_no_footprints_not_near_promoter,\n",
    "                                  logFC=edge_logFC,logFC_p_value=logFC_p_value_)]\n",
    "    idx <- idx+1\n",
    "}\n",
    "p_value_dt[,motif_p_adj:=p.adjust(motif_p_value) ]\n",
    "frac_occ_dt[,`:=`(fold_enrichment=promoter_frac/non_promoter_frac,logFC_p_adj=p.adjust(logFC_p_value))]\n",
    "frac_occ_dt <- merge( frac_occ_dt, p_value_dt )\n",
    "frac_occ_dt[,tf:=gsub(\"MA..*\\\\.([A-Z])\",\"\\\\1\",motif)]\n",
    "\n",
    "sig_frac_occ_dt <- frac_occ_dt[motif_p_adj < 0.05,]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Edge cells in other datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GSE81547"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_dir <- \"/Users/sreenivasagopv2/Data/sc-funnel/GSE81547/\"\n",
    "file_paths = list.files(gene_exp_dir,full.names=T)\n",
    "gene_exp_dt <- data.table()\n",
    "col_idx <- 1\n",
    "for (file_path in file_paths) {\n",
    "    if (grepl(\".gz\",file_path)) {\n",
    "        print(col_idx)\n",
    "        flush.console()\n",
    "        dt <- fread( file_path )\n",
    "        gsm_num <- str_match( file_path, \"GSM[0-9][0-9]*\" )[1]\n",
    "        if (nrow(gene_exp_dt) == 0 ) {\n",
    "            gene_exp_dt <- dt %>% setnames(.,\"V2\",gsm_num)\n",
    "        } else {\n",
    "            gene_exp_dt <- cbind( gene_exp_dt, dt[,list(V2)] ) %>% setnames(.,\"V2\",gsm_num)\n",
    "        }\n",
    "        col_idx <- col_idx + 1\n",
    "    }\n",
    "}\n",
    "fwrite(gene_exp_dt,out_path,quote=F,row.names=F,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path <- \"/Users/sreenivasagopv2/Data/sc-funnel/GSE81547/GSE81547_mat.tsv.gz\"\n",
    "\n",
    "gene_exp_mat <- read_gene_exp_mat(out_path)\n",
    "aging_obj <- create_full_seurat_object( gene_exp_mat )\n",
    "aging_obj <- subset( aging_obj, features=rownames(seurat_obj[[\"RNA\"]]@counts) )\n",
    "aging_obj <- NormalizeData( aging_obj )\n",
    "\n",
    "rm(gene_exp_mat)\n",
    "\n",
    "geo_info <- getGEO(\"GSE81547\",getGPL=F)\n",
    "geo_info_df <- pData(geo_info$GSE81547_series_matrix.txt.gz)\n",
    "donor_age <- str_match(geo_info_df[,\"title\"],\"^[0-9][0-9]*\")\n",
    "info_to_add_df <- data.frame( age=as.integer(donor_age), row.names=Cells(aging_obj))\n",
    "\n",
    "anno_dt <- fread(file.path(base_path,\"acinar_scrna_seq.txt\"))\n",
    "cell_types <- anno_dt[match(`Sample Name`,rownames(info_to_add_df)),inferred_cell_type]\n",
    "info_to_add_df$cell_type <- cell_types\n",
    "\n",
    "aging_obj <- AddMetaData( aging_obj, info_to_add_df )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aging_obj <- SetIdent(aging_obj,value=\"cell_type\")\n",
    "seurat_obj <- SetIdent(seurat_obj,value=\"cluster\")\n",
    "aging_acinar_cells <- WhichCells(aging_obj,idents=\"acinar\")\n",
    "aging_acinar_data_mat <- aging_obj[[\"RNA\"]]@data[,aging_acinar_cells]\n",
    "\n",
    "all_edge_acinar_genes <- diff_exp_dt_list[[\"Acinar cell\"]]$gene_name\n",
    "all_edge_acinar_genes <- all_edge_acinar_genes[all_edge_acinar_genes %in% rownames(aging_acinar_data_mat)]\n",
    "original_edge_acinar_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                                           cell_category == \"edge\",cell.name]\n",
    "original_non_edge_acinar_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                                           cell_category == \"center\",cell.name]\n",
    "original_acinar_cells <- c(original_edge_acinar_cells,original_non_edge_acinar_cells)\n",
    "original_acinar_obj <- subset( seurat_obj, idents=\"Acinar cell\")# %>% ScaleData(features=all_edge_acinar_genes)\n",
    "acinar_data_mat <- as.matrix(original_acinar_obj[[\"RNA\"]]@data[all_edge_acinar_genes,original_acinar_cells])\n",
    "aging_acinar_data_mat <- aging_acinar_data_mat[all_edge_acinar_genes,]\n",
    "\n",
    "rm(original_acinar_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_cos_sim_mat <- get_cosine_matrix(acinar_data_mat,acinar_data_mat)\n",
    "query_cos_sim_mat <- get_cosine_matrix(aging_acinar_data_mat,acinar_data_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_mat <- function(disp_mat,num_clusters) {\n",
    "     pathway_hclust_obj <- hclust( dist(disp_mat) )\n",
    "#     pathway_hclust_obj <- hclust( dist(disp_mat), method=\"complete\" )\n",
    "\n",
    "    pathway_clusters <- cutree(tree = pathway_hclust_obj, k = num_clusters)\n",
    "    cluster_colors <- list()\n",
    "    cluster_colors[[\"Cluster\"]] <- brewer.pal(n = num_clusters, name = \"Accent\")\n",
    "    names(cluster_colors[[\"Cluster\"]]) <- paste(\"Cluster\", 1:num_clusters)\n",
    "\n",
    "    pathway_cluster_dt <- data.table( cluster=paste( \"Cluster\", pathway_clusters ) )\n",
    "    rownames(pathway_cluster_dt) <- names(pathway_clusters)\n",
    "    \n",
    "    return(pathway_cluster_dt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "debug_cos_sim_mat <- get_cosine_matrix(aging_acinar_data_mat,aging_acinar_data_mat)\n",
    "\n",
    "options(repr.plot.width=12,repr.plot.height=12)\n",
    "age_dt <- cluster_mat(debug_cos_sim_mat,6)\n",
    "pheatmap( debug_cos_sim_mat, show_colnames=F, \n",
    "         annotation_row=age_dt, show_rownames=F, fontsize=15 )\n",
    "age_dt$cell.name <- rownames(age_dt)\n",
    "age_dt <- merge( age_dt, aging_meta_data_dt[,.(cell.name,age)] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_sim_dt <- data.table(reshape2::melt(reference_cos_sim_mat[original_edge_acinar_cells,\n",
    "                                                              original_edge_acinar_cells],\n",
    "                                        value.name=\"sim\"))[Var1 != Var2,]\n",
    "median_edge_sim_dt <- edge_sim_dt[,median_sim:=median(sim),by=\"Var1\"][,.(cell.name=Var1,median_sim)] %>% unique(.)\n",
    "non_edge_sim_dt <- data.table(reshape2::melt(reference_cos_sim_mat[original_non_edge_acinar_cells,\n",
    "                                                              original_non_edge_acinar_cells],\n",
    "                                        value.name=\"sim\"))[Var1 != Var2,]\n",
    "median_non_edge_sim_dt <- non_edge_sim_dt[,median_sim:=median(sim),by=\"Var1\"][,.(cell.name=Var1,median_sim)] %>% unique(.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_dt <- data.table(reshape2::melt(query_cos_sim_mat))\n",
    "query_dt[Var2 %in% original_edge_acinar_cells,cell_type:=\"Edge\"]\n",
    "query_dt[!Var2 %in% original_edge_acinar_cells,cell_type:=\"Non-Edge\"]\n",
    "query_dt[,median_sim:=median(value),by=list(Var1,cell_type)]\n",
    "query_dt <- merge( query_dt, aging_meta_data_dt[,.(cell.name,age)],by.x=\"Var1\",by.y=\"cell.name\")\n",
    "\n",
    "options(repr.plot.height=10,repr.plot.width=10)\n",
    "plot_dt <- unique(query_dt[,.(Var1,median_sim,cell_type,age)] )\n",
    "plot_dt$age <- factor(plot_dt$age,levels=sort(unique(plot_dt$age)))\n",
    "ggplot( plot_dt ) + \n",
    "       geom_boxplot(aes(x=age,y=median_sim,color=cell_type),size=1) + theme_classic(base_size=20) +\n",
    "theme(panel.grid.major=element_line(color=\"gray\",linetype=\"dotted\"),legend.position=\"bottom\") + \n",
    "ylab(\"Median Similarity to Edge Acinar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_prob <- kde(x=median_edge_sim_dt$median_sim,eval.points=query_median_sim_dt[cell_type==\"Edge\",median_sim])$estimate\n",
    "non_edge_prob <- kde(x=median_non_edge_sim_dt$median_sim,eval.points=query_median_sim_dt[cell_type!=\"Edge\",median_sim])$estimate\n",
    "likelihood_dt <- data.table( cell.name=unique(query_median_sim_dt$Var1), ratio=edge_prob/non_edge_prob, inferred_type=\"Non-Edge\" )\n",
    "likelihood_dt <- merge( likelihood_dt, aging_meta_data_dt[,.(cell.name,age)] )\n",
    "likelihood_dt[ratio > 1.5,inferred_type:=\"Edge\"]\n",
    "temp_dt <- likelihood_dt[,.N,by=list(age,inferred_type)][order(age),]\n",
    "edge_dt <- temp_dt[inferred_type == \"Edge\",.(age,N_edge=N)]\n",
    "total_dt <- likelihood_dt[,.N,by=\"age\"][,.(age,N_total=N)]\n",
    "merged_dt <- merge( edge_dt, total_dt )[,frac_edge:=N_edge/N_total]\n",
    "print(merged_dt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Human cell atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_dt <- fread(file.path(base_path,\"MCA_CellAssignments.csv\"),header=T,sep=\",\")\n",
    "adult_pancreas_df <- as.data.frame( anno_dt[Tissue == \"Pancreas\",.(cell.name=gsub(\"Pancreas_1\\\\.\",\"\",Cell.name),\n",
    "                                                                   cell.type=gsub(\"\\\\(Pancreas\\\\)\",\"\",Annotation))])\n",
    "rownames(adult_pancreas_df) <- adult_pancreas_df$cell.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_mat <- read_gene_exp_mat(file.path(base_path,\"GSM4008637_Adult-Pancreas1_dge.txt.gz\"))\n",
    "hcl_adult_obj <- create_full_seurat_object( gene_exp_mat )\n",
    "# hcl_adult_obj <- subset( hcl_adult_obj, cells=adult_pancreas_df$cell.name )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hcl_adult_obj[[\"percent.mt\"]] <- PercentageFeatureSet(hcl_adult_obj, pattern = \"^MT-\")\n",
    "hcl_meta_data_dt <- data.table( hcl_adult_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "non_dying_cells <- hcl_meta_data_dt[percent.mt < 10,cell.name]\n",
    "hcl_adult_obj <- subset( hcl_adult_obj, cells=non_dying_cells)\n",
    "hcl_meta_data_dt <- data.table( hcl_adult_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hcl_adult_obj <- NormalizeData(hcl_adult_obj) %>% ScaleData(.) %>% \n",
    "    FindVariableFeatures(.,nfeatures = 500) %>% RunPCA(.,dims=30) %>% FindNeighbors( ., reduction = \"pca\" ) %>%\n",
    "    FindClusters(.) %>% RunUMAP(.,dims=1:30)\n",
    "hcl_meta_data_dt <- data.table( hcl_adult_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hcl_adult_obj <- FindClusters(hcl_adult_obj,resolution=0.1)\n",
    "options(repr.plot.width=10,repr.plot.height=10)\n",
    "theme_set(theme_classic(base_size=20))\n",
    "DimPlot( hcl_adult_obj, reduction=\"umap\", group.by=\"seurat_clusters\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c(\"PRSS1\",\"CTRB1\",\"CTRB2\",\"REG1B\")\n",
    "temp_df <- FetchData( hcl_adult_obj, vars=)\n",
    "marker_sum <- rowSums(temp_df)\n",
    "temp_dt <- data.table( cell.name=names(marker_sum), marker_score=marker_sum ) %>% \n",
    "merge(.,hcl_meta_data_dt[,.(cell.name,seurat_clusters)],by=\"cell.name\")\n",
    "#PRSS1, CTRB1, CTRB2, REG1B\n",
    "# FeaturePlot( hcl_adult_obj, features=c(\"CTRB1\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "score_dt <- temp_dt[,mean_score:=mean(marker_score),by=seurat_clusters] %>% unique(.,by=\"seurat_clusters\")\n",
    "score_dt[order(-mean_score),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GSE85241"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_mat <- read_gene_exp_mat(file.path(base_path,\"GSE85241_cellsystems_dataset_4donors_updated.csv.gz\"))\n",
    "\n",
    "processed_gene_names <- gsub(\"__.*\",\"\",rownames(gene_exp_mat))\n",
    "names(processed_gene_names) <- rownames(gene_exp_mat)\n",
    "duplicate_gene_names <- processed_gene_names[duplicated( processed_gene_names ) == T]\n",
    "unique_gene_names <- processed_gene_names[!processed_gene_names %in% duplicate_gene_names]\n",
    "new_gene_exp_mat <- gene_exp_mat[names(unique_gene_names),]\n",
    "rownames(new_gene_exp_mat) <- unique_gene_names\n",
    "new_mat <- matrix(0,nrow=length(duplicate_gene_names),ncol=ncol(new_gene_exp_mat),\n",
    "                 dimnames=list(duplicate_gene_names,colnames(new_gene_exp_mat)))\n",
    "for (gene in duplicate_gene_names) {\n",
    "    entries <- rownames(gene_exp_mat)[grepl(gene,rownames(gene_exp_mat))]\n",
    "    new_mat[gene,] <- rowSums(gene_exp_mat[entries,])\n",
    "}\n",
    "new_gene_exp_mat <- rbind(new_gene_exp_mat,new_mat)\n",
    "cel_seq_obj <- create_full_seurat_object( new_gene_exp_mat )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cel_seq_obj <- subset( cel_seq_obj, subset=`nFeature_RNA` > 1000)\n",
    "# cel_seq_meta_data_dt <- data.table( cel_seq_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "# viable_cells <- cel_seq_meta_data_dt[nFeature_RNA > 200,cell.name]\n",
    "cel_seq_obj <- NormalizeData(cel_seq_obj) %>% ScaleData(.) %>% \n",
    "    FindVariableFeatures(.,nfeatures = 2000) %>% RunPCA(.,dims=50) %>% FindNeighbors( ., reduction = \"pca\", k.param=30 ) %>%\n",
    "    FindClusters(.) %>% RunUMAP(.,dims=1:50)\n",
    "cel_seq_meta_data_dt <- data.table( cel_seq_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "prss1_dt <- FetchData( cel_seq_obj, vars=c(\"seurat_clusters\",\"PRSS1\")) %>% data.table(.)\n",
    "prss1_dt[,mean_prss1:=mean(PRSS1),by=\"seurat_clusters\"]\n",
    "prss1_dt <- prss1_dt[order(-mean_prss1),.(seurat_clusters,mean_prss1)] %>% unique(.) #Cluster 2 corresponds to acinar cells\n",
    "acinar_cells <- cel_seq_meta_data_dt[seurat_clusters == prss1_dt[1,seurat_clusters],cell.name]\n",
    "\n",
    "\n",
    "age_meta_data_dt <- data.table( donor_id=c(\"D28\",\"D29\",\"D30\",\"D31\",\"D16\",\"D25\"),\n",
    "                              age=c(54,23,48,59,53,30),\n",
    "                              sex=c(\"M\",\"M\",\"F\",\"M\",\"M\",\"M\"))\n",
    "\n",
    "cel_seq_meta_data_dt$donor_id <- gsub(\"-[0-9][0-9]*\",\"\",cel_seq_meta_data_dt$orig.ident)\n",
    "\n",
    "cel_seq_meta_data_dt <- merge( cel_seq_meta_data_dt, age_meta_data_dt, by=\"donor_id\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acinar_obj <- subset( cel_seq_obj, cells=acinar_cells ) %>% NormalizeData(.) %>% ScaleData(.,features=rownames(new_gene_exp_mat))\n",
    "# acinar_obj <- subset( cel_seq_obj, cells=acinar_cells ) %>% SCTransform(.) %>% RunPCA(.,dims=50)\n",
    "acinar_obj <- FindVariableFeatures(acinar_obj,nfeatures = 1000) %>% RunPCA(.,dims=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj <- SetIdent(seurat_obj,value=\"cluster\")\n",
    "other_acinar_data_mat <- acinar_obj[[\"RNA\"]]@data\n",
    "\n",
    "all_edge_acinar_genes <- diff_exp_dt_list[[\"Acinar cell\"]]$gene_name\n",
    "all_edge_acinar_genes <- all_edge_acinar_genes[all_edge_acinar_genes %in% rownames(other_acinar_data_mat)]\n",
    "original_edge_acinar_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                                           cell_category == \"edge\",cell.name]\n",
    "original_non_edge_acinar_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                                           cell_category == \"center\",cell.name]\n",
    "original_acinar_cells <- c(original_edge_acinar_cells,original_non_edge_acinar_cells)\n",
    "original_acinar_obj <- subset( seurat_obj, idents=\"Acinar cell\")# %>% ScaleData(features=all_edge_acinar_genes)\n",
    "acinar_data_mat <- as.matrix(original_acinar_obj[[\"RNA\"]]@data[all_edge_acinar_genes,original_acinar_cells])\n",
    "other_acinar_data_mat <- other_acinar_data_mat[all_edge_acinar_genes,]\n",
    "\n",
    "rm(original_acinar_obj)\n",
    "reference_cos_sim_mat <- get_cosine_matrix(acinar_data_mat,acinar_data_mat)\n",
    "query_cos_sim_mat <- get_cosine_matrix(other_acinar_data_mat,acinar_data_mat)\n",
    "\n",
    "query_dt <- data.table(reshape2::melt(query_cos_sim_mat))\n",
    "query_dt[Var2 %in% original_edge_acinar_cells,cell_type:=\"Edge\"]\n",
    "query_dt[!Var2 %in% original_edge_acinar_cells,cell_type:=\"Non-Edge\"]\n",
    "query_dt[,median_sim:=median(value),by=list(Var1,cell_type)]\n",
    "# query_dt <- merge( query_dt, aging_meta_data_dt[,.(cell.name,age)],by.x=\"Var1\",by.y=\"cell.name\")\n",
    "\n",
    "pca_distances <- self_pca_distances( acinar_cells, Embeddings(acinar_obj,reduction=\"pca\"))\n",
    "pca_dist_dt <- data.table( cell.name=names(pca_distances), dist_from_medoid=pca_distances )\n",
    "query_dt <- merge(query_dt,pca_dist_dt,by.x=\"Var1\",by.y=\"cell.name\")\n",
    "plot_dt <- query_dt[,.(Var1,cell_type,median_sim,dist_from_medoid)] %>% unique(.) %>% \n",
    "merge(.,cel_seq_meta_data_dt[,.(cell.name,age,sex)],by.x=\"Var1\",by.y=\"cell.name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( plot_dt ) + geom_boxplot(aes(x=as.character(age),y=median_sim,color=cell_type))  +\n",
    "theme_classic(base_size=20) + ylab(\"Cosine similarity\") + xlab(\"Age\") + theme(legend.position = \"bottom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( plot_dt ) + geom_point(aes(x=dist_from_medoid,y=median_sim)) + facet_wrap(~cell_type) + scale_x_log10()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_sim_dt <- data.table(reshape2::melt(reference_cos_sim_mat[original_edge_acinar_cells,\n",
    "                                                              original_edge_acinar_cells],\n",
    "                                        value.name=\"sim\"))[Var1 != Var2,]\n",
    "median_edge_sim_dt <- edge_sim_dt[,median_sim:=median(sim),by=\"Var1\"][,.(cell.name=Var1,median_sim)] %>% unique(.)\n",
    "non_edge_sim_dt <- data.table(reshape2::melt(reference_cos_sim_mat[original_non_edge_acinar_cells,\n",
    "                                                              original_non_edge_acinar_cells],\n",
    "                                        value.name=\"sim\"))[Var1 != Var2,]\n",
    "\n",
    "median_non_edge_sim_dt <- non_edge_sim_dt[,median_sim:=median(sim),by=\"Var1\"][,.(cell.name=Var1,median_sim)] %>% unique(.)\n",
    "edge_prob <- kde(x=median_edge_sim_dt$median_sim,eval.points=plot_dt[cell_type==\"Edge\",median_sim])$estimate\n",
    "non_edge_prob <- kde(x=median_non_edge_sim_dt$median_sim,eval.points=plot_dt[cell_type!=\"Edge\",median_sim])$estimate\n",
    "likelihood_dt <- data.table( cell.name=unique(plot_dt$Var1), ratio=edge_prob/non_edge_prob, inferred_type=\"Non-Edge\" )\n",
    "likelihood_dt <- merge( likelihood_dt, cel_seq_meta_data_dt[,.(cell.name,age)] )\n",
    "likelihood_dt[ratio > 1.5,inferred_type:=\"Edge\"]\n",
    "temp_dt <- likelihood_dt[,.N,by=list(age,inferred_type)][order(age),]\n",
    "edge_dt <- temp_dt[inferred_type == \"Edge\",.(age,N_edge=N)]\n",
    "total_dt <- likelihood_dt[,.N,by=\"age\"][,.(age,N_total=N)]\n",
    "merged_dt <- merge( edge_dt, total_dt )[,frac_edge:=N_edge/N_total]\n",
    "print(merged_dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "324.2px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
