{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘robustbase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘ggpubr’ was built under R version 3.6.2”\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "Loading required package: VGAM\n",
      "\n",
      "Loading required package: splines\n",
      "\n",
      "Loading required package: DDRTree\n",
      "\n",
      "Loading required package: irlba\n",
      "\n",
      "Loading required package: Rcpp\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "library(Seurat)\n",
    "#library(SeuratData)\n",
    "library(ggplot2)\n",
    "source(\"../src/process_pdac.R\")\n",
    "source(\"../src/pdac_plots.R\")\n",
    "#library(parallelDist)\n",
    "library(grid)\n",
    "library(dplyr)\n",
    "library(pheatmap)\n",
    "library(DoubletFinder)\n",
    "library(ggpubr)\n",
    "library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(BSgenome.Hsapiens.UCSC.hg19)\n",
    "library(robustbase)\n",
    "library(gridExtra)\n",
    "library(RColorBrewer)\n",
    "library(stringr)\n",
    "library(monocle)\n",
    "library(fgsea)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PDAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in fread(matrix_path):\n",
      "“Detected 57530 column names but the data has 57531 columns (i.e. invalid file). Added 1 extra default column name for the first column which is guessed to be row names or an index. Use setnames() afterwards if this guess is not correct, or fix the file write command that created the file to create a valid file.”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n"
     ]
    }
   ],
   "source": [
    "base_path <- file.path(\"..\",\"data\")\n",
    "pdac_matrix_file_name <- \"CRA001160.matrix\"\n",
    "pdac_anno_file_name <- \"CRA001160.celltypes.tsv\"\n",
    "pdac_matrix_path <- file.path( base_path, pdac_matrix_file_name )\n",
    "pdac_annotation_path <- file.path( base_path, pdac_anno_file_name )\n",
    "\n",
    "pdac_anno_dt <- process_annotations( pdac_annotation_path )\n",
    "seurat_obj <- create_full_seurat_object( read_gene_exp_mat( \n",
    "    pdac_matrix_path ), pdac_anno_dt )\n",
    "meta_data_dt <- data.table( seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "\n",
    "doublet_cells <- fread( file.path( base_path, \"cell_doublets.tsv\" ),header=F) %>% .$V1\n",
    "\n",
    "filtered_cells <- meta_data_dt[!cell.name %in% doublet_cells,cell.name]\n",
    "pdac_anno_dt <- pdac_anno_dt[cell.name %in% filtered_cells,]\n",
    "\n",
    "seurat_obj <- subset( seurat_obj, cells=filtered_cells )\n",
    "seurat_obj <- NormalizeData( seurat_obj )\n",
    "meta_data_dt <- data.table( seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "edge_info_all <- readRDS( file.path( base_path, \"edge_info_all_doublet_filtered_celltypes.rds\" ) )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison with trajectory analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_obj <- readRDS(file.path(base_path,\"ductal_malignant_cds_obj.rds\"))\n",
    "cell_type <- \"Ductal cell type 1\"\n",
    "cds_meta_data_dt <- data.table( pData(cds_obj) )\n",
    "cells <- cds_meta_data_dt[!cell.name %in% doublet_cells,cell.name]\n",
    "# cds_obj <- cds_obj[,cds_meta_data_dt$cell.name]\n",
    "cds_meta_data_dt$sub_cluster <- cds_meta_data_dt$cluster\n",
    "cds_meta_data_dt[cell.name %in% edge_info_all$edge_center_dt[normal_cell_type==cell_type & cell_category == \"edge\",cell.name],sub_cluster:=paste(\"Edge\", cell_type)]\n",
    "cds_meta_data_dt[!grepl(\"Edge\",sub_cluster) & cluster == cell_type,sub_cluster:=paste(\"Non-Edge\", cell_type)]\n",
    "df <- as.data.frame(cds_meta_data_dt[,!c(\"cell.name\")])\n",
    "rownames(df) <- cds_meta_data_dt$cell.name\n",
    "pData(cds_obj) <- df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=14,repr.plot.height=6)\n",
    "p1 <- plot_cell_trajectory( cds_obj, color_by=\"sub_cluster\" ) + labs(color=\"Cell Type\") + theme_classic(base_size=15) +\n",
    "theme(legend.position=c(1,0.8),legend.justification = \"right\",legend.text=element_text(size=15)) +\n",
    "scale_color_discrete(name=\"Cell Type\",labels=c(\"Ductal cell type 2\"=\"Malignant Ductal\",\n",
    "                                              \"Edge Ductal cell type 1\"=\"Outlier Ductal\",\n",
    "                                              \"Non-Edge Ductal cell type 1\"=\"Non-Outlier Ductal\"))\n",
    "\n",
    "p2 <- ggboxplot( cds_meta_data_dt[sub_cluster!=\"Acinar cell\",], x=\"State\",y=\"Pseudotime\",color=\"sub_cluster\" ) + \n",
    "theme_classic(base_size=15) + xlab(\"State\") +\n",
    "stat_compare_means(comparisons=list(c(\"Non-Edge Acinar cell\",\"Edge Acinar cell\")),label.y=7) +\n",
    "theme(legend.position=\"none\")\n",
    "final <- ggarrange(p1,p2,nrow=1)\n",
    "print(final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=14,repr.plot.height=6)\n",
    "p1 <- plot_cell_trajectory( cds_obj, color_by=\"sub_cluster\" ) + labs(color=\"\") + theme_classic(base_size=15) +\n",
    "theme(legend.position=c(1,0.8),legend.justification = \"right\",legend.text=element_text(size=15)) #+ \n",
    "# scale_color_manual(breaks=c(\"Ductal cell type 2\"=\"Malignant Ductal\",\"Edge Ductal Cell Type 1\"=\"Outlier Ductal\",\n",
    "#                            \"Non-Edge Ductal cell type 1\"=\"Non-Outlier Ductal\"))\n",
    "\n",
    "\n",
    "p2 <- ggboxplot( cds_meta_data_dt[sub_cluster!=cell_type,], x=\"sub_cluster\",y=\"Pseudotime\" ) + \n",
    "theme_classic(base_size=15) + xlab(\"\") +\n",
    "stat_compare_means(comparisons=list(c(\"Non-Edge Ductal cell type 1\",\"Edge Ductal cell type 1\")),label.y=20)\n",
    "final <- ggarrange(p1,p2,nrow=1)\n",
    "print(final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10,repr.plot.height=6)\n",
    "# p1 <- plot_cell_trajectory( cds_obj, color_by=\"sub_cluster\" ) + labs(color=\"\") + theme_classic(base_size=15) + \n",
    "# facet_grid(rows=vars(sub_cluster),cols=vars(State)) + theme(legend.position=\"bottom\",legend.justification = \"right\",legend.text=element_text(size=15)) #+ \n",
    "# # scale_color_manual(breaks=c(\"Ductal cell type 2\"=\"Malignant Ductal\",\"Edge Ductal Cell Type 1\"=\"Outlier Ductal\",\n",
    "# #                            \"Non-Edge Ductal cell type 1\"=\"Non-Outlier Ductal\"))\n",
    "\n",
    "\n",
    "p2 <- ggboxplot( cds_meta_data_dt, color=\"sub_cluster\",y=\"Pseudotime\",x=\"State\" ) + \n",
    "theme_classic(base_size=20) + labs(color=\"Cell Type\") +\n",
    "stat_compare_means(comparisons=list(c(\"Non-Edge Ductal cell type 1\",\"Edge Ductal cell type 1\"))) + \n",
    "theme(legend.position=\"bottom\",legend.text=element_text(size=15))\n",
    "p2\n",
    "# final <- ggarrange(p1,p2,ncol=1)\n",
    "# print(final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sub-sampling reads and genes from edge cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"process_pdac.R\")\n",
    "source(\"pdac_plots.R\")\n",
    "norm_type <- \"Acinar cell\"\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=3)\n",
    "#theme_set(theme_classic(base_size = 15))\n",
    "resample_info_list <- resample_edge_cells( seurat_obj, pdac_anno_dt, edge_info_all$edge_center_dt, c(\"Acinar cell\"), c(\"Ductal cell type 2\"), \"reads\",\n",
    "                                      no_tumour_adjacent=F)\n",
    "\n",
    "#Before resampling\n",
    "p1 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes(x=cell_category_before_resampling,y=log(nCount_RNA_before_resampling))) + \n",
    "stat_compare_means(aes(x=cell_category_before_resampling,y=log(nCount_RNA_before_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() +\n",
    "xlab(\"Acinar cell category\") + ylab(\"Log(Library size)\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\n",
    "                                                                                    \"edge\"=\"Edge\"))\n",
    "\n",
    "#After resampling\n",
    "p2 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes(x=cell_category_after_resampling,y=log(nCount_RNA_after_resampling))) +\n",
    "stat_compare_means(aes(x=cell_category_after_resampling,y=log(nCount_RNA_after_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() + ylab(\"\") +\n",
    " xlab(\"Acinar cell category\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\"edge\"=\"Edge\"))\n",
    " \n",
    "p3 <- plot_skewness(resample_info_list$edge_info$control_dist_dt) + theme_classic()\n",
    "\n",
    "p4 <- plot_edge_distance_ratio(resample_info_list$edge_info$edge_malignant_dist_dt) + theme_classic()\n",
    "                              \n",
    "final <- ggarrange(p1,p2,p3,p4,nrow=1,ncol=4,labels=\"auto\")\n",
    "print(final)\n",
    "\n",
    "num_retained_edge_cells <- nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\" & \n",
    "                                         cell_category_after_resampling == \"edge\",] )\n",
    "print(paste(\"# edge cells =\",nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\",])))\n",
    "print(paste(\"# common edge cells before and after resampling =\",num_retained_edge_cells))\n",
    "ggsave(\"supp_library_resampling.png\", final, height=3, width=8 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"process_pdac.R\")\n",
    "source(\"pdac_plots.R\")\n",
    "norm_type <- \"Acinar cell\"\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=3)\n",
    "#theme_set(theme_classic(base_size = 15))\n",
    "resample_info_list <- resample_edge_cells( seurat_obj, pdac_anno_dt, edge_info_all$edge_center_dt, c(\"Acinar cell\"), \n",
    "                                           c(\"Ductal cell type 2\"), \"genes\", no_tumour_adjacent=F)\n",
    "\n",
    "#Before resampling\n",
    "p1 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes(x=cell_category_before_resampling,y=log(nFeature_RNA_before_resampling))) + \n",
    "stat_compare_means(aes(x=cell_category_before_resampling,y=log(nFeature_RNA_before_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() +\n",
    "xlab(\"Acinar cell category\") + ylab(\"Log(Library size)\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\n",
    "                                                                                    \"edge\"=\"Edge\"))\n",
    "\n",
    "#After resampling\n",
    "p2 <- ggplot( resample_info_list$resampled_meta_data_dt ) + geom_boxplot(aes (x=cell_category_after_resampling,y=log(nFeature_RNA_after_resampling))) +\n",
    "stat_compare_means(aes(x=cell_category_after_resampling,y=log(nFeature_RNA_after_resampling)),label=\"p.format\",label.x=1.5) + theme_classic() + ylab(\"\") +\n",
    "xlab(\"Acinar cell category\") + scale_x_discrete(labels=c(\"center\"=\"Non-edge\",\"edge\"=\"Edge\"))\n",
    " \n",
    "p3 <- plot_skewness(resample_info_list$edge_info$control_dist_dt) + theme_classic()\n",
    "\n",
    "p4 <- plot_edge_distance_ratio(resample_info_list$edge_info$edge_malignant_dist_dt) + theme_classic()\n",
    "                              \n",
    "final <- ggarrange(p1,p2,p3,p4,nrow=1,ncol=4,labels=\"auto\")\n",
    "print(final)\n",
    "\n",
    "num_retained_edge_cells <- nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\" & \n",
    "                                         cell_category_after_resampling == \"edge\",] )\n",
    "print(paste(\"# edge cells =\",nrow(resample_info_list$resampled_meta_data_dt[cell_category_before_resampling == \"edge\",])))\n",
    "print(paste(\"# common edge cells before and after resampling =\",num_retained_edge_cells))\n",
    "ggsave(\"supp_genes_resampling.png\", final, height=3, width=8 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Justifying merging normal and tumour-adjacent acinar cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_type <- \"Acinar cell\"\n",
    "tumour_adjacent_cells <- meta_data_dt[cluster == norm_type & sample_type == \"tumour\",cell.name]\n",
    "\n",
    "# acinar_edge_info <- add_edge_center_annotation( seurat_obj, pdac_anno_dt, \n",
    "#                                                       malignant_cell_types=c(\"Ductal cell type 2\"), \n",
    "#                                   normal_cell_types=c(norm_type), num_pcs=num_pcs, \n",
    "#                                   pairwise_pca=T, perform_control=F, no_tumour_adjacent=T )\n",
    "\n",
    "cell_type_acinar_edge_center_dt <- acinar_edge_info$edge_center_dt\n",
    "diff_exp_dt <- get_edge_de_genes_dt( seurat_obj, cell_type_acinar_edge_center_dt, \n",
    "                                    tumour_adjacent_cells = tumour_adjacent_cells )\n",
    "diff_exp_dt <- diff_exp_dt[p_val_adj < 0.01,]\n",
    "\n",
    "acinar_edge_cells <- cell_type_acinar_edge_center_dt[cell_category == \"edge\",cell.name]\n",
    "non_acinar_edge_cells <- cell_type_acinar_edge_center_dt[cell_category != \"edge\",cell.name]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=3)\n",
    "theme_set(theme_classic())\n",
    "\n",
    "merged_diff_exp_dt <- merge( diff_exp_dt[de_set == \"tumour_adjacent_vs_center\"],\n",
    "                         diff_exp_dt[de_set == \"edge_vs_center\"], by=\"gene_name\",\n",
    "                           suffixes=c(\".adj_vs_center\",\".edge_vs_center\"))\n",
    "p1 <- ggplot( merged_diff_exp_dt ) + geom_point(aes(x=avg_logFC.edge_vs_center,y=avg_logFC.adj_vs_center)) + \n",
    "geom_line(aes(x=avg_logFC.edge_vs_center,y=avg_logFC.edge_vs_center),color=\"gray\") + \n",
    "geom_hline(aes(yintercept=0)) + geom_vline(aes(xintercept=0)) + xlab(\"Average Log FC\\n(Edge vs Non-Edge)\") +\n",
    "ylab(\"Average Log FC\\n(Tumor-adjacent vs Non-Edge)\")\n",
    "print(paste(\"R^2\"=cor.test(merged_diff_exp_dt$avg_logFC.edge_vs_center,merged_diff_exp_dt$avg_logFC.adj_vs_center)))\n",
    "\n",
    "plot_dt <- melt( merged_diff_exp_dt[,.(gene_name,avg_logFC.edge_vs_center,avg_logFC.adj_vs_center)],\n",
    "                id.vars=\"gene_name\", value.name=\"Average Log FC\")\n",
    "p2 <- ggplot( plot_dt ) + geom_point(aes(x=variable,y=`Average Log FC`),position=\"jitter\",color=\"gray\") +\n",
    "geom_boxplot(aes(x=variable,y=`Average Log FC`),outlier.shape=NA,fill=NA,size=1.0) + \n",
    "scale_x_discrete(\"\",labels=c(\"avg_logFC.edge_vs_center\" = \"Edge vs\\n Non-Edge\",\n",
    "                          \"avg_logFC.adj_vs_center\" = \"Tumor-Adjacent vs\\n Non-Edge\")) + \n",
    "stat_compare_means(aes(x=variable,y=`Average Log FC`),label.x=1.5,label=\"p.signif\")\n",
    "\n",
    "acinar_edge_center_dt <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\",.(cell.name,dist_from_normal_medoid,dist_from_malignant_medoid)]\n",
    "acinar_edge_center_dt <- merge( acinar_edge_center_dt, meta_data_dt[,.(cell.name,sample_type)] )\n",
    "melted_dt <- melt(acinar_edge_center_dt[sample_type==\"tumour\",][,!c(\"sample_type\")],id.vars=c(\"cell.name\"))\n",
    "p3 <- ggplot( melted_dt ) + stat_compare_means(aes(x=variable,y=value),label=\"..p.format..\",method.args=list(\"alternative\"=\"g\"),label.x=1.5) +\n",
    " geom_boxplot(aes(x=variable,y=value)) + scale_x_discrete(c(\"variable\"=\"\"), labels=c(\"dist_from_normal_medoid\"=\"From\\nacinar\",\n",
    "                                                           \"dist_from_malignant_medoid\"=\"From\\nmalignant ductal\")) + ylab(\"Distance in PC space\")\n",
    "ggarrange(p1,p2,p3ncol=3,nrow=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(paste0(\"# of DE genes (edge vs center) \", nrow(diff_exp_dt[de_set == \"edge_vs_center\"])))\n",
    "print(paste0(\"# of DE genes (tumour-adjacent vs center) \", nrow(diff_exp_dt[de_set == \"tumour_adjacent_vs_center\"])))\n",
    "print(paste0(\"# of common DE genes \", nrow(merged_diff_exp_dt)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Edge-ness fingerprint plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cell_types <- unique(pdac_anno_dt$cluster)\n",
    "across_refs_edge_center_dt <- data.table()\n",
    "num_pcs = 50\n",
    "\n",
    "malignant_cell_types <- c(\"Ductal cell type 2\")\n",
    "normal_cell_types <- setdiff( unique(pdac_anno_dt$cluster), malignant_cell_types)\n",
    "edge_info_all <- add_edge_center_annotation( seurat_obj, pdac_anno_dt, malignant_cell_types=malignant_cell_types, \n",
    "                                  normal_cell_types=normal_cell_types, num_pcs=num_pcs, \n",
    "                                  pairwise_pca=T, perform_control=T, no_tumour_adjacent=F )\n",
    "\n",
    "saveRDS(edge_info_all,\"edge_info_all_doublet_filtered_celltypes.rds\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison with embryonic progenitor genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n"
     ]
    }
   ],
   "source": [
    "diff_exp_dt_list <- list()\n",
    "\n",
    "edge_non_edge_zscores_list <- list()\n",
    "edge_mean_exp_list <- list()\n",
    "non_edge_mean_exp_list <- list()\n",
    "for (cell_type in c(\"Acinar cell\",\"Ductal cell type 1\")) {\n",
    "    edge_center_dt <- edge_info_all$edge_center_dt[normal_cell_type == cell_type,]\n",
    "\n",
    "    diff_exp_dt_list[[cell_type]] <- get_edge_de_genes_dt( seurat_obj, \n",
    "                                          edge_center_dt                \n",
    "                                        )[p_val_adj < 0.01,]\n",
    "    subset_seurat_obj <- subset( seurat_obj, cells=edge_center_dt$cell.name ) %>% ScaleData(.)\n",
    "    edge_mean_exp_list[[cell_type]] <- rowMeans(subset_seurat_obj[[\"RNA\"]]@data[,edge_center_dt[cell_category == \"edge\",cell.name]])\n",
    "    non_edge_mean_exp_list[[cell_type]] <- rowMeans(subset_seurat_obj[[\"RNA\"]]@data[,edge_center_dt[cell_category == \"center\",cell.name]])\n",
    "\n",
    "    edge_non_edge_zscores <- rowMeans(subset_seurat_obj[[\"RNA\"]]@scale.data[,edge_center_dt[cell_category == \"edge\",cell.name]]) -\n",
    "    rowMeans(subset_seurat_obj[[\"RNA\"]]@scale.data[,edge_center_dt[cell_category == \"center\",cell.name]])\n",
    "    \n",
    "    edge_non_edge_zscores_list[[cell_type]] <- edge_non_edge_zscores\n",
    "}\n",
    "rm(subset_seurat_obj)\n",
    "\n",
    "derived_signatures <- list()\n",
    "\n",
    "common_edge_genes <- intersect( diff_exp_dt_list[[\"Acinar cell\"]]$gene_name, \n",
    "                               diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name )\n",
    "acinar_only_edge_genes <- setdiff( diff_exp_dt_list[[\"Acinar cell\"]]$gene_name, common_edge_genes )\n",
    "ductal_only_edge_genes <- setdiff( diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name, common_edge_genes )\n",
    "all_edge_genes <- union( diff_exp_dt_list[[\"Acinar cell\"]]$gene_name, \n",
    "                               diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name )\n",
    "\n",
    "derived_signatures[[\"Edge Acinar All\"]] <- diff_exp_dt_list[[\"Acinar cell\"]]$gene_name\n",
    "derived_signatures[[\"Edge Acinar Unique\"]] <- acinar_only_edge_genes\n",
    "derived_signatures[[\"Outlier Ductal All\"]] <- diff_exp_dt_list[[\"Ductal cell type 1\"]]$gene_name\n",
    "derived_signatures[[\"Outlier Ductal Unique\"]] <- ductal_only_edge_genes\n",
    "derived_signatures[[\"Outlier Universe\"]] <- all_edge_genes\n",
    "derived_signatures[[\"Edge Outlier Common\"]] <- common_edge_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_table <- fread(\"progenitor_marker_table.tsv\")[,!c(\"Stainings\\ and\\ References\")]\n",
    "set(marker_table,NULL,\"Acinar -> Progenitor\",paste(marker_table$`Acinar cell`, marker_table$`Embryonic progenitor`,sep=\" to \"))\n",
    "set(marker_table,NULL,\"Ductal -> Progenitor\",paste(marker_table$`Duct cell`, marker_table$`Embryonic progenitor`,sep=\" to \"))\n",
    "marker_table <- marker_table[,!c(\"Acinar cell\",\"Duct cell\",\"Embryonic progenitor\",\"\")]\n",
    "marker_genes <- marker_table$Marker\n",
    "\n",
    "new_markers <- c(\"STAT3\",\"SEL1L\",\"CBL\",\"KLF4\",\"CTNND1\",\"LKB1\",\"ICAM1\",\"DCLK1\",\"CDKN1A\")\n",
    "marker_genes <- unique( c(marker_genes,new_markers) )\n",
    "marker_genes <- marker_genes[marker_genes %in% names(edge_non_edge_zscores_list[[\"Acinar cell\"]])]\n",
    "\n",
    "for (gene in marker_genes) {\n",
    "    if (!gene %in% marker_table$Marker){\n",
    "        marker_table <- rbind( marker_table, list(gene,\"- to +\",\"- to +\"))\n",
    "    }\n",
    "}\n",
    "rows_to_update <- 1:nrow(marker_table)\n",
    "for (cell_type in c(\"Acinar cell\",\"Ductal cell type 1\")) {\n",
    "    de_status_vec <- rep(\"\",length(marker_genes))#vector(mode=\"character\",length=length(marker_genes))\n",
    "    names(de_status_vec) <- marker_genes\n",
    "    edge_column_values <- rep(\"N/A\",length(marker_genes))\n",
    "    names(edge_column_values) <- marker_genes\n",
    "\n",
    "    de_marker_genes <- marker_genes[marker_genes %in% diff_exp_dt_list[[cell_type]]$gene_name]\n",
    "    de_status_vec[de_marker_genes] <- \"*\"\n",
    "    edge_column_values[marker_genes] <- paste(round(edge_non_edge_zscores_list[[cell_type]][marker_genes],2),de_status_vec)\n",
    "    #edge_column_values[is.na(edge_column_values)] <- \"N/A\"\n",
    "    edge_column_name <- paste(gsub(\"\\ .*\",\"\",cell_type),\"Non-Edge -> Edge\")\n",
    "    \n",
    "    set(marker_table,i=rows_to_update[marker_table$Marker %in% marker_genes],\n",
    "        edge_column_name,edge_column_values)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "progenitor_count_mat <- read_gene_exp_mat(\"/Users/sreenivasagopv2/Data/sc-funnel/GSM4194789_TMM_counts_CPM.csv.gz\")\n",
    "\n",
    "if (!file.exists(\"progenitor_annotation.tsv\")) {\n",
    "    progenitor_seurat_obj <- CreateSeuratObject( progenitor_count_mat )\n",
    "    progenitor_seurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(progenitor_seurat_obj, pattern = \"^MT-\")\n",
    "    dying_cells <- WhichCells( progenitor_seurat_obj, expression='percent.mt > 10' )\n",
    "    \n",
    "    progenitor_seurat_obj <- NormalizeData(progenitor_seurat_obj) \n",
    "    progenitor_seurat_obj <- ScaleData(progenitor_seurat_obj) %>% FindVariableFeatures(.,nfeatures = 1000) %>% RunPCA(.,dims=50) %>% RunUMAP(.,dims=1:50) \n",
    "\n",
    "    homotypic.prop <- modelHomotypic(progenitor_seurat_obj@meta.data$cluster)\n",
    "    nExp_poi <- round(0.05*nrow(progenitor_seurat_obj@meta.data))\n",
    "    #nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop)) \n",
    "\n",
    "    sample_param_sweep <- paramSweep_v3(progenitor_seurat_obj, PCs = 1:50, sct = FALSE)\n",
    "    sample_param_sweep_summary <- summarizeSweep(sample_param_sweep, GT = FALSE)\n",
    "    sweep_dt <- data.table( find.pK( sample_param_sweep_summary ) )\n",
    "    optimal_pK <- as.double(as.vector(sweep_dt[order(-BCmetric),][1]$pK))\n",
    "    progenitor_seurat_obj <- doubletFinder_v3(progenitor_seurat_obj, \n",
    "                               PCs = 1:50, pN = 0.25, pK = optimal_pK, nExp = nExp_poi, \n",
    "                                          reuse.pANN = FALSE, sct = FALSE)\n",
    "\n",
    "    sample_meta_data_dt <- data.table( progenitor_seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "    doublet_class_col <- paste(\"DF.classifications_0.25\",optimal_pK,nExp_poi,sep=\"_\")\n",
    "    doublet_info_dt <- sample_meta_data_dt[,c(\"cell.name\",doublet_class_col),with=F] %>%  setnames(.,doublet_class_col,\"doublet_class\")\n",
    "\n",
    "    cell_doublets <- progenitor_meta_data_dt[cell.name %in% doublet_info_dt[doublet_class == \"Doublet\",cell.name],cell.name]\n",
    "    #fwrite( data.table(cell.name=cell_doublets), \"progenitor_cell_doublets.tsv\", sep=\"\\t\", col.names=F, quote=F)\n",
    "    \n",
    "    cells_to_retain <- setdiff( Cells(progenitor_seurat_obj), c(dying_cells,doublet_cells) )\n",
    "    progenitor_seurat_obj <- subset( progenitor_seurat_obj, cells=cells_to_retain )\n",
    "    progenitor_meta_data_dt <- data.table( progenitor_seurat_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "\n",
    "    progenitor_seurat_obj <- NormalizeData(progenitor_seurat_obj) %>% ScaleData(.) %>% \n",
    "    FindVariableFeatures(.,nfeatures = 1000) %>% RunPCA(.,dims=50) %>% FindNeighbors( ., reduction = \"pca\", k.param=30 ) %>%\n",
    "    FindClusters(.) %>% RunUMAP(.,dims=1:50)\n",
    "    progenitor_meta_data_dt <- data.table( progenitor_seurat_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "    fwrite( progenitor_meta_data_dt, \"progenitor_annotation.tsv\", sep=\"\\t\", row.names=F,quote=F)\n",
    "} else {\n",
    "    progenitor_meta_data_dt <- fread(\"progenitor_annotation.tsv\")\n",
    "    progenitor_seurat_obj <- create_full_seurat_object( progenitor_count_mat, progenitor_meta_data_dt )\n",
    "    progenitor_seurat_obj <- NormalizeData( progenitor_seurat_obj )\n",
    "    #Cluster 7 contains Sox9+,PTF1a+ cells\n",
    "    sox9_ptf1a_cells <- progenitor_meta_data_dt[seurat_clusters == 7,cell.name]\n",
    "}\n",
    "\n",
    "\n",
    "progenitor_seurat_obj <- SetIdent( progenitor_seurat_obj, value=\"seurat_clusters\" )\n",
    "progenitor_markers_df <- FindMarkers( progenitor_seurat_obj, ident.1=\"7\")\n",
    "progenitor_markers_dt <- data.table( progenitor_markers_df, keep.rownames=T ) %>% setnames(.,\"rn\",\"gene_name\")\n",
    "progenitor_genes <- progenitor_markers_dt[p_val < 0.1 & avg_logFC > 0,gene_name]\n",
    "progenitor_genes <- progenitor_genes[progenitor_genes %in% all_genes]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cosine_matrix <- function(X,Y) {\n",
    "    norm_X <- sqrt(colSums(X^2))\n",
    "#     print(Y)\n",
    "    norm_Y <- sqrt(colSums(Y^2))\n",
    "    \n",
    "    dimnames_X <- list(rownames(X),colnames(X))\n",
    "    dimnames_Y <- list(rownames(Y),colnames(Y))\n",
    "\n",
    "    X <- sapply(1:ncol(X),function(idx){return(X[,idx]/norm_X[idx])})\n",
    "    Y <- sapply(1:ncol(Y),function(idx){return(Y[,idx]/norm_Y[idx])})\n",
    "    \n",
    "    dimnames(X) <- dimnames_X\n",
    "    dimnames(Y) <- dimnames_Y\n",
    "    \n",
    "    dist_mat <- t(X) %*% Y\n",
    "    return(dist_mat)\n",
    "}\n",
    "\n",
    "compute_cell_wise_cosine_similarity <- function( query_mat, reference_mat, genes_to_use=NULL, \n",
    "                                                dist_from=\"medoid\",return_sd=F ) {\n",
    "    if (!is.null(genes_to_use)) {\n",
    "        query_mat <- query_mat[genes_to_use,]\n",
    "        reference_mat <- reference_mat[genes_to_use,]\n",
    "    } \n",
    "    \n",
    "    query_ref_dist_mat <- get_cosine_matrix( query_mat, reference_mat )\n",
    "    \n",
    "    if (dist_from == \"medoid\") {\n",
    "        ref_dist_mat <- get_cosine_matrix( reference_mat, reference_mat )\n",
    "        medoid <-which.max(rowSums(ref_dist_mat))\n",
    "        distances <- query_ref_dist_mat[,medoid]\n",
    "        to_return <- list(\"dist\"=distances)\n",
    "    } else if (dist_from == \"all\") {\n",
    "        distances <- Matrix::rowMeans(query_ref_dist_mat)\n",
    "        num_ref_cells <- ncol(reference_mat)\n",
    "        sd_distances <- sqrt( (num_ref_cells/(num_ref_cells-1)) * (rowMeans(query_ref_dist_mat^2) - distances^2))\n",
    "        to_return <- list(\"dist\"=distances,\"sd\"=sd_distances)\n",
    "    }\n",
    "    \n",
    "    return(to_return)\n",
    "}\n",
    "\n",
    "plot_similarity_scatter <- function( dist_list_x, dist_list_y ) {\n",
    "    dt <- data.table( x_mean=dist_list_x$dist,\n",
    "                     y_mean=dist_list_y$dist,\n",
    "                      x_sd=dist_list_x$sd,\n",
    "                     y_sd=dist_list_y$sd )\n",
    "    \n",
    "    p <- ggplot( dt ) + geom_pointrange(aes(x=x_mean,y=y_mean,xmin=x_mean-1.96*x_sd,\n",
    "                                            xmax=x_mean+1.96*x_sd),fatten=1,color=\"gray\",\n",
    "                                        size=0.5) +\n",
    "       geom_pointrange(aes(x=x_mean,y=y_mean,ymin=y_mean-1.96*y_sd,ymax=y_mean+1.96*y_sd),fatten=1,color=\"gray\",\n",
    "                       size=0.5) + \n",
    "    geom_abline(slope=1,intercept=0) + geom_point(aes(x=x_mean,y=y_mean),color=\"black\",size=1) +\n",
    "    xlim(0,1) + ylim(0,1)\n",
    " \n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "progenitor_mat <- progenitor_seurat_obj[[\"RNA\"]]@data[,sox9_ptf1a_cells]\n",
    "\n",
    "acinar_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category==\"edge\",cell.name]\n",
    "acinar_non_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category!=\"edge\",cell.name]\n",
    "\n",
    "ductal_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                     cell_category == \"edge\",cell.name]\n",
    "ductal_non_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                     cell_category != \"edge\",cell.name]\n",
    "\n",
    "malignant_ductal_cells <- meta_data_dt[cluster == \"Ductal cell type 2\",cell.name]\n",
    "\n",
    "# acinar_seurat_obj <- \n",
    "ductal_outlier_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,ductal_outlier_cells]\n",
    "acinar_edge_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,acinar_edge_cells]\n",
    "ductal_non_outlier_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,ductal_non_outlier_cells]\n",
    "acinar_non_edge_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,acinar_non_edge_cells]\n",
    "malignant_data_mat <- seurat_obj[[\"RNA\"]]@data[progenitor_genes,malignant_ductal_cells]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ret_edge_acinar <- compute_cell_wise_cosine_similarity( acinar_edge_data_mat, progenitor_mat,\n",
    "                                           genes_to_use = progenitor_genes, dist_from=\"medoid\")\n",
    "ret_outlier_ductal <- compute_cell_wise_cosine_similarity( ductal_outlier_data_mat, progenitor_mat,\n",
    "                                           genes_to_use = progenitor_genes, dist_from=\"medoid\")\n",
    "\n",
    "ret_non_edge_acinar <- compute_cell_wise_cosine_similarity( acinar_non_edge_data_mat,progenitor_mat,\n",
    "                                           genes_to_use = progenitor_genes, dist_from=\"medoid\")\n",
    "ret_non_outlier_ductal <- compute_cell_wise_cosine_similarity( ductal_non_outlier_data_mat, progenitor_mat,\n",
    "                                           genes_to_use = progenitor_genes, dist_from=\"medoid\")\n",
    "\n",
    "ret_malignant <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, progenitor_mat,\n",
    "                                            genes_to_use = progenitor_genes, dist_from=\"medoid\")\n",
    "\n",
    "edge_acinar_dist_dt <- data.table(value=ret_edge_acinar$dist,variable=\"Edge Acinar\")\n",
    "outlier_ductal_dist_dt <- data.table(value=ret_outlier_ductal$dist,variable=\"Outlier Ductal\")\n",
    "non_edge_acinar_dist_dt <- data.table(value=ret_non_edge_acinar$dist,variable=\"Non-Edge Acinar\")\n",
    "non_outlier_ductal_dist_dt <- data.table(value=ret_non_outlier_ductal$dist,variable=\"Non-Outlier Ductal\")\n",
    "malignant_dist_dt <- data.table(value=ret_malignant$dist,variable=\"Malignant Ductal\")\n",
    "\n",
    "\n",
    "options(repr.plot.width=4,repr.plot.height=6)\n",
    "plot_dt <- rbind(edge_acinar_dist_dt,non_edge_acinar_dist_dt,outlier_ductal_dist_dt,\n",
    "                 non_outlier_ductal_dist_dt,malignant_dist_dt)\n",
    "\n",
    "plot_dt$cell_type <- plot_dt$variable\n",
    "plot_dt[grepl(\"Acinar\",variable),]$cell_type <- \"Acinar\"\n",
    "plot_dt[grepl(\"Ductal cell type 1\",variable),]$cell_type <- \"Ductal\"\n",
    "plot_dt[grepl(\"Ductal cell type 2\",variable),]$cell_type <- \"Ductal\"\n",
    "\n",
    "ggboxplot(plot_dt,x=\"variable\",y=\"value\",color=\"cell_type\") +\n",
    "theme_classic(base_size=15) + ylab(\"Cosine similarity\") + xlab(\"\") + \n",
    "# theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) +\n",
    "stat_compare_means(comparisons=list(c(\"Edge Acinar\",\"Outlier Ductal\"),\n",
    "                                    c(\"Edge Acinar\",\"Non-Edge Acinar\"),\n",
    "                                   c(\"Outlier Ductal\",\"Non-Outlier Ductal\"),\n",
    "                                   c(\"Malignant Ductal\",\"Edge Acinar\")),step.increase=0.2,label.x=1.5,label=\"p.signif\",size=8) +\n",
    "scale_x_discrete(labels=c(\"Edge Acinar\"=\"E\",\"Outlier Ductal\"=\"O\",\"Non-Edge Acinar\"=\"NE\",\"Non-Outlier Ductal\"=\"NO\")) + \n",
    "labs(color=\"Cell Type\") + theme(legend.position=\"bottom\") + ylim(0,1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6,repr.plot.height=8)\n",
    "plot_dt <- rbind(edge_acinar_dist_dt,non_edge_acinar_dist_dt,outlier_ductal_dist_dt,\n",
    "                 non_outlier_ductal_dist_dt,malignant_dist_dt)\n",
    "\n",
    "plot_dt$cell_type <- plot_dt$variable\n",
    "plot_dt[grepl(\"Acinar\",variable),]$cell_type <- \"Acinar\"\n",
    "plot_dt[grepl(\"Ductal cell type 1\",variable),]$cell_type <- \"Ductal\"\n",
    "plot_dt[grepl(\"Ductal cell type 2\",variable),]$cell_type <- \"Malignant\"\n",
    "\n",
    "ggboxplot(plot_dt,x=\"variable\",y=\"value\",color=\"cell_type\") +\n",
    "theme_classic(base_size=15) + ylab(\"Cosine similarity\") + xlab(\"\") + \n",
    "# theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) +\n",
    "stat_compare_means(comparisons=list(c(\"Edge Acinar\",\"Outlier Ductal\"),\n",
    "                                    c(\"Edge Acinar\",\"Non-Edge Acinar\"),\n",
    "                                   c(\"Outlier Ductal\",\"Non-Outlier Ductal\"),\n",
    "                                   c(\"Malignant Ductal\",\"Edge Acinar\")),label.x=1.5,label=\"p.signif\",size=8) +\n",
    "scale_x_discrete(labels=c(\"Edge Acinar\"=\"E\",\"Outlier Ductal\"=\"O\",\"Non-Edge Acinar\"=\"NE\",\n",
    "                          \"Non-Outlier Ductal\"=\"NO\",\"Malignant Ductal\"=\"Mal\")) + \n",
    "labs(color=\"Cell Type\") + theme(legend.position=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FGSEA enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_h <- gmtPathways(file.path(base_path,\"h.all.v7.0.symbols.gmt\"))\n",
    "\n",
    "all_genes <- rownames(seurat_obj[[\"RNA\"]]@data)\n",
    "\n",
    "cancersea_gene_sets <- list()\n",
    "signature_file_paths <- list.files(file.path(base_path,\"cancer-sea\"),full.names=T)\n",
    "for (idx in 1:length(signature_file_paths)) {\n",
    "    signature_genes_dt <- fread(signature_file_paths[idx])\n",
    "    temp <- unlist(strsplit(signature_file_paths[idx],\"\\\\.\"))[1] \n",
    "    signature_name <- paste( \"CancerSEA\", tail( unlist( strsplit( temp, \"/\" ) ), 1 ), sep=\"_\" )\n",
    "    signature_gene_names <- signature_genes_dt$GeneName\n",
    "    cancersea_gene_sets[[signature_name]] <- signature_gene_names\n",
    "}\n",
    "pathways_all <- c(pathways_h,cancersea_gene_sets)\n",
    "\n",
    "num_pathways <- length(pathways_all)\n",
    "\n",
    "pathways_all <- lapply( pathways_all, function(pathway_genes) {return(pathway_genes[pathway_genes %in% all_genes])})\n",
    "for (pathway in names(pathways_all)) {\n",
    "    if (length(pathways_all[[pathway]]) == 0) {\n",
    "        pathways_all[[pathway]] <- NULL\n",
    "    }\n",
    "} \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conventional GSEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normal_cell_types <- c(\"Acinar cell\",\"Ductal cell type 1\")\n",
    "malignant_cell_type <- \"Ductal cell type 2\"\n",
    "malignant_cells <- meta_data_dt[cluster == malignant_cell_type,cell.name]\n",
    "fgsea_dt <- data.table(pathway=names(pathways_all))\n",
    "\n",
    "comparison_list <- list(\"Malignant v All Acinar\"=c(\"Ductal cell type 2\",\"Acinar cell\"),\n",
    "                       \"Edge v Non-Edge Acinar\"=c(\"edge Acinar cell\",\"center Acinar cell\"),\n",
    "                        \"Malignant v All Ductal\"=c(\"Ductal cell type 2\",\"Ductal cell type 1\"),\n",
    "                       \"Outlier v Non-Outlier Ductal\"=c(\"edge Ductal cell type 1\",\"center Ductal cell type 1\"))\n",
    "\n",
    "get_cells <- function(cluster_name) {\n",
    "    if (!grepl(\"edge\",cluster_name) && !grepl(\"center\",cluster_name)) {\n",
    "        cells <- meta_data_dt[cluster == cluster_name,cell.name]\n",
    "    } else {\n",
    "        if (grepl(\"edge\",cluster_name)) {\n",
    "            cell_category_ <- \"edge\"\n",
    "        } else {\n",
    "            cell_category_ <- \"center\"\n",
    "        }\n",
    "        cluster_name <- gsub(paste(cell_category_,\"\"),\"\",cluster_name)\n",
    "        cells <- edge_info_all$edge_center_dt[normal_cell_type == cluster_name & cell_category == cell_category_,\n",
    "                                             cell.name]\n",
    "    }\n",
    "    \n",
    "    return(cells)\n",
    "}\n",
    "\n",
    "\n",
    "for (comparison in names(comparison_list)) {\n",
    "    print(comparison)\n",
    "    flush.console()\n",
    "    column_pair <- comparison_list[[comparison]]\n",
    "    cells_1 <- get_cells( column_pair[1] )\n",
    "    cells_2 <- get_cells( column_pair[2] )\n",
    "    cells <- union(cells_1,cells_2)\n",
    "    subset_seurat_obj <- subset( seurat_obj, cells=cells ) %>% ScaleData(.)\n",
    "    scaled_data <- subset_seurat_obj[[\"RNA\"]]@scale.data\n",
    "    \n",
    "    zscore_diff <- rowMeans(scaled_data[,cells_1]) - rowMeans(scaled_data[,cells_2])\n",
    "\n",
    "    dt <- run_fgsea( pathways_all, zscore_diff)\n",
    "    fgsea_dt <- merge( fgsea_dt, dt[,.(pathway,padj,NES)],by=\"pathway\") %>% \n",
    "    setnames(.,\"padj\",paste(\"padj\",comparison)) %>% setnames(.,\"NES\",paste(\"NES\",comparison))\n",
    "    rm(scaled_data)\n",
    "    rm(subset_seurat_obj)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "make_pathway_annotation <- function(disp_mat,num_clusters) {\n",
    "     pathway_hclust_obj <- hclust( dist(t(scale(t(disp_mat)))) )\n",
    "#     pathway_hclust_obj <- hclust( dist(disp_mat), method=\"complete\" )\n",
    "\n",
    "    pathway_clusters <- cutree(tree = pathway_hclust_obj, k = num_clusters)\n",
    "    cluster_colors <- list()\n",
    "    cluster_colors[[\"Cluster\"]] <- brewer.pal(n = num_clusters, name = \"Accent\")\n",
    "    names(cluster_colors[[\"Cluster\"]]) <- paste(\"Cluster\", 1:num_clusters)\n",
    "\n",
    "    pathway_cluster_dt <- data.table( `Gene Set`=paste( \"Cluster\", pathway_clusters ) )\n",
    "    rownames(pathway_cluster_dt) <- names(pathway_clusters)\n",
    "    \n",
    "    return(pathway_cluster_dt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_sig <- function(col) {\n",
    "    row_mask <- fgsea_dt[[paste(\"padj\", col)]] < 0.05\n",
    "    #idxes <- 1:nrow(fgsea_dt)\n",
    "    return(row_mask)\n",
    "}\n",
    "cluster_dt <- data.table(pathway=fgsea_dt$pathway,`Enriched in`=\"C8:Misc.\")\n",
    "\n",
    "cluster_dt[get_sig(\"Edge v Non-Edge Acinar\"),`Enriched in`:=\"C6:EA-NEA\"]\n",
    "cluster_dt[get_sig(\"Outlier v Non-Outlier Ductal\"),`Enriched in`:=\"C7:OD-NOD\"]\n",
    "cluster_dt[get_sig(\"Malignant v All Acinar\"),`Enriched in`:=\"C3:Mal-Aci\"]\n",
    "cluster_dt[get_sig(\"Malignant v All Ductal\"),`Enriched in`:=\"C4:Mal-Duc\"]\n",
    "\n",
    "\n",
    "cluster_dt[get_sig(\"Malignant v All Ductal\") & get_sig(\"Malignant v All Acinar\"),\n",
    "                   `Enriched in`:=\"C2:Mal-Duc & Mal-Aci\"]\n",
    "cluster_dt[get_sig(\"Edge v Non-Edge Acinar\") & get_sig(\"Outlier v Non-Outlier Ductal\"),\n",
    "           `Enriched in`:=\"C5:EA-NEA & OD-NOD\"]\n",
    "cluster_dt[get_sig(\"Malignant v All Ductal\") & get_sig(\"Malignant v All Acinar\") &\n",
    "                get_sig(\"Edge v Non-Edge Acinar\") & get_sig(\"Outlier v Non-Outlier Ductal\"),\n",
    "                   `Enriched in`:=\"C1:All\"]\n",
    "\n",
    "cluster_dt <- cluster_dt[order(`Enriched in`),]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "row_order <- c(\"C1:All\",\"C2:Mal-Duc & Mal-Aci\",\n",
    "                                        \"C3:Mal-Aci\",\n",
    "                                        \"C4:Mal-Duc\",\n",
    "                                       \"C5:EA-NEA & OD-NOD\",\n",
    "                                       \"C6:EA-NEA\",\n",
    "                                       \"C7:OD-NOD\")\n",
    "                                         #\"C8:Misc.\")\n",
    "pathway_cluster_dt <- data.table()\n",
    "gap_rows <- c()\n",
    "idx <- 0\n",
    "new_row_entries <- c()\n",
    "for (row in row_order) {\n",
    "    if (row == \"Misc.\")\n",
    "        next\n",
    "    dt <- cluster_dt[`Enriched in` == row,]\n",
    "    dt$`Enriched in` <- paste(row, paste0(\"(\",nrow(dt),\")\"))\n",
    "    pathway_cluster_dt <- rbind( pathway_cluster_dt, dt)\n",
    "    idx <- idx + nrow(dt)\n",
    "    gap_rows <- c(gap_rows,idx)\n",
    "    new_row_entries <- c(new_row_entries,dt$`Enriched in`)\n",
    "}\n",
    "pathway_cluster_dt$`Enriched in` <- factor(pathway_cluster_dt$`Enriched in`,levels=unique(new_row_entries))\n",
    "pathway_order <- pathway_cluster_dt$pathway\n",
    "pathway_cluster_dt <- pathway_cluster_dt[,!c(\"pathway\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "theme_set(theme_gray(base_size = 8))\n",
    "\n",
    "all_col_names <- names(fgsea_dt)\n",
    "nes_cols <- all_col_names[grepl(\"NES\",all_col_names)]\n",
    "padj_cols <- all_col_names[grepl(\"padj\",all_col_names)]\n",
    "\n",
    "# sig_pathways <- melt(fgsea_dt[,c(\"pathway\",padj_cols),with=F],id.vars=\"pathway\")[value < 0.05,pathway] %>% unique(.)\n",
    "fgsea_dt <- fgsea_dt[order(match(pathway,pathway_order)),]\n",
    "pathway_mat <- as.matrix(fgsea_dt[,nes_cols,with=F])\n",
    "colnames(pathway_mat) <- gsub(\"NES \",\"\", nes_cols )\n",
    "rownames(pathway_mat) <- fgsea_dt$pathway\n",
    "\n",
    "flip_to_NA <- function(x) {\n",
    "    nes_scores <- pathway_mat[,x]\n",
    "    p_vals <- fgsea_dt[[paste(\"padj\",x)]]\n",
    "    nes_scores[p_vals > 0.05] = NA\n",
    "#      nes_scores[nes_scores > 0] <- 1\n",
    "#      nes_scores[nes_scores < 0] <- -1\n",
    "     return(nes_scores)\n",
    "}\n",
    "\n",
    "na_pathway_mat <- sapply( names(comparison_list), flip_to_NA )\n",
    "colnames(na_pathway_mat) <- colnames(pathway_mat)\n",
    "rownames(pathway_cluster_dt) <- pathway_order\n",
    "na_pathway_mat <- na_pathway_mat[pathway_order,]\n",
    "p <- pheatmap(na_pathway_mat,fontsize=15,cluster_cols=F,cluster_rows=F,show_rownames=T,\n",
    "         annotation_row=pathway_cluster_dt, na_col=\"gray\",border_color=\"black\",\n",
    "              colorRampPalette(rev(brewer.pal(n = 8, name = \"RdYlBu\")))(4),\n",
    "             breaks=c(-2,-1,0,1,2),\n",
    "              gaps_row=gap_rows,cellheight=15,cellwidth=15,\n",
    "             filename=\"enrichment_heatmap.png\",width=20,height=20,dpi=300)\n",
    "             "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tumour-wise scoring of ductal and acinar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acinar_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category==\"edge\",cell.name]\n",
    "acinar_non_edge_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & cell_category!=\"edge\",cell.name]\n",
    "\n",
    "ductal_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                     cell_category != \"edge\",cell.name]\n",
    "ductal_non_outlier_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Ductal cell type 1\" & \n",
    "                                                      cell_category != \"edge\",cell.name]\n",
    "\n",
    "ductal_non_outlier_data_mat <- seurat_obj[[\"RNA\"]]@data[,ductal_non_outlier_cells]\n",
    "acinar_non_edge_data_mat <- seurat_obj[[\"RNA\"]]@data[,acinar_non_edge_cells]\n",
    "\n",
    "ret_edge_acinar <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, acinar_edge_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Edge Acinar All\"]], dist_from=\"medoid\")\n",
    "ret_outlier_ductal <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, ductal_outlier_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Outlier Ductal All\"]], dist_from=\"medoid\")\n",
    "ret_non_edge_acinar <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, acinar_non_edge_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Edge Acinar All\"]], dist_from=\"medoid\")\n",
    "ret_non_outlier_ductal <- compute_cell_wise_cosine_similarity( malignant_ductal_mat, ductal_non_outlier_data_mat,\n",
    "                                           genes_to_use=derived_signatures[[\"Outlier Ductal All\"]], dist_from=\"medoid\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=16,repr.plot.height=6)\n",
    "melted_dt <- melt(data.table(`Edge Acinar`=ret_edge_acinar$dist,\n",
    "                        `Outlier Ductal`=ret_non_outlier_ductal$dist,\n",
    "                        `Non-Edge Acinar`=ret_non_edge_acinar$dist,\n",
    "                        `Non-Outlier Ductal`=ret_non_outlier_ductal$dist,\n",
    "                       cell.name=names(ret_edge_acinar$dist)),id.vars=c(\"cell.name\")\n",
    "                 ) #%>% setnames(.,\"variable\",\"Cell Type\")\n",
    "# 4 + \"\"\n",
    "melted_dt <- merge( melted_dt, meta_data_dt[,.(sample,cell.name)])\n",
    "melted_dt$sample <- factor(melted_dt$sample,levels=paste0(\"T\",1:24))\n",
    "\n",
    "p_edge_outlier <- ggplot( melted_dt[!grepl(\"Non-\",variable),]  ) + \n",
    "geom_boxplot(aes(y=value,x=sample,color=variable),size=1.0,fill=NA,width=0.5) + \n",
    "# stat_compare_means(aes(x=variable,y=value),label.x=1.5,label=\"p.signif\",size=8) + \n",
    "xlab(\"\") + ylab(\"\") +\n",
    "theme_classic(base_size=20) + labs(color=\"\") + theme(legend.position=c(0.5,0.1), legend.direction=\"horizontal\")\n",
    "\n",
    "x_labels <- 1:24\n",
    "names(x_labels) <- paste0(\"T\",x_labels)\n",
    "p_all <- ggplot( melted_dt  ) + \n",
    "geom_boxplot(aes(y=value,x=sample,color=variable),size=1.0,fill=NA,width=0.5) + \n",
    "theme_classic(base_size=20) + ylab(\"Cosine similarity\") + xlab(\"\") + labs(color=\"Cell Type\") +\n",
    "ylim(0.3,1) + theme(legend.direction=\"horizontal\",legend.position=c(0.5,0.1))\n",
    "\n",
    "\n",
    "melted_dt$cell_type <- melted_dt$variable\n",
    "melted_dt[grepl(\"Acinar\",variable),]$cell_type <- \"Acinar\"\n",
    "melted_dt[grepl(\"Ductal\",variable),]$cell_type <- \"Ductal\"\n",
    "# ggplot( melted_dt ) + \n",
    "# geom_boxplot(aes(y=value,group=variable),size=1.0,fill=NA,width=0.5) + \n",
    "p_pan_tumor <- ggboxplot(melted_dt,x=\"variable\",y=\"value\",color=\"cell_type\") +\n",
    "theme_classic(base_size=20) + ylab(\"Cosine similarity\") + xlab(\"\") + \n",
    "# theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) +\n",
    "stat_compare_means(comparisons=list(c(\"Edge Acinar\",\"Outlier Ductal\"),\n",
    "                                    c(\"Edge Acinar\",\"Non-Edge Acinar\"),\n",
    "                                   c(\"Outlier Ductal\",\"Non-Outlier Ductal\")),\n",
    "                   label.x=1.5,label=\"p.signif\",size=5) +\n",
    "scale_x_discrete(labels=c(\"Edge Acinar\"=\"E\",\"Outlier Ductal\"=\"O\",\"Non-Edge Acinar\"=\"NE\",\"Non-Outlier Ductal\"=\"NO\")) + \n",
    "scale_fill_discrete(name = c(\"cell_type\"=\"Cell Type\")) + labs(color=\"\") + \n",
    "theme(legend.position=\"none\") #+ ylim(0.3,0.9)\n",
    "\n",
    "print(\"# of tumors where edge acinar is closer to malignant than outlier ductal\")\n",
    "p_values <- c()\n",
    "for (sample_num in unique(melted_dt$sample)) {\n",
    "    dt <- melted_dt[sample == sample_num,]\n",
    "    p_value <- wilcox.test(dt[variable == \"Edge Acinar\",value],dt[variable == \"Outlier Ductal\",value])$p.value\n",
    "    p_values <- c(p_values,p_value)\n",
    "}\n",
    "adj_p <- p.adjust(p_values,method=\"bonferroni\")\n",
    "print(sum(adj_p < 0.1))\n",
    "\n",
    "rm(melted_dt)\n",
    "rm(ductal_outlier_data_mat)\n",
    "rm(ductal_non_outlier_data_mat)\n",
    "rm(acinar_edge_data_mat)\n",
    "rm(acinar_non_edge_data_mat)\n",
    "\n",
    "final <- ggarrange(p_pan_tumor,p_edge_outlier,nrow=1,ncol=2,widths=c(0.3,1),labels=\"auto\")\n",
    "print(final)\n",
    "print(p_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mechanisms of edge-ness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_footprints_dt <- fread(file.path(base_path,\"match\",\"footprints_mpbs.bed\"))[,!c(\"V6\",\"V7\")] %>%\n",
    "setnames(.,paste0(\"V\",1:5),c(\"chr\",\"start\",\"end\",\"motif\",\"motif_score\")) %>% unique(.)\n",
    "motif_granges <- makeGRangesFromDataFrame( motif_footprints_dt, keep.extra.columns=T )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned many:many mapping between keys and columns\n",
      "\n",
      "Warning message in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE):\n",
      "“GRanges object contains 1 out-of-bound range located on sequence\n",
      "  chrUn_gl000223. Note that ranges located on a sequence whose length is\n",
      "  unknown (NA) or on a circular sequence are not considered out-of-bound\n",
      "  (use seqlengths() and isCircular() to get the lengths and circularity\n",
      "  flags of the underlying sequences). You can use trim() to trim these\n",
      "  ranges. See ?`trim,GenomicRanges-method` for more information.”\n"
     ]
    }
   ],
   "source": [
    "# all_genes <- rownames(seurat_obj[[\"RNA\"]]@data)\n",
    "# entrez_dt <- get_entrez_dt( all_genes )\n",
    "entrez_ids <- entrez_dt$final_entrez_id\n",
    "acinar_edge_up_genes <- diff_exp_dt_list[[\"Acinar cell\"]][avg_logFC > 0,gene_name]\n",
    "entrez_dt[SYMBOL %in% acinar_edge_up_genes,`:=`(gene_type=\"Edge\")]\n",
    "entrez_dt[is.na(gene_type),gene_type:=\"Not Edge\"]\n",
    "cds_tx_dt <- select(TxDb.Hsapiens.UCSC.hg19.knownGene,columns=c(\"GENEID\",\"TXNAME\"),\n",
    "               keytype=c(\"GENEID\"), keys=entrez_ids ) %>% na.omit(.)\n",
    "cds_tx_dt <- merge( cds_tx_dt, entrez_dt[,.(SYMBOL,final_entrez_id,gene_type)],\n",
    "                   by.x=\"GENEID\", by.y=\"final_entrez_id\") %>% data.table(.)\n",
    "\n",
    "promoters_txdb <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene,upstream=2000,downstream=400) %>% trim(.)\n",
    "\n",
    "mcol_df <- as.data.frame(mcols(promoters_txdb))\n",
    "mcol_df$gene_type <- \"Not Edge\"\n",
    "mcol_df[mcol_df$tx_name %in% cds_tx_dt[gene_type == \"Edge\",TXNAME],]$gene_type <- \"Edge\"\n",
    "mcols(promoters_txdb) <- mcol_df\n",
    "rm(mcol_df) \n",
    "\n",
    "overlap_dt <- data.table( as.data.frame(findOverlaps( motif_granges, promoters_txdb ) ) )\n",
    "promoter_dt <- data.table( as.data.frame(promoters_txdb))\n",
    "\n",
    "overlap_motif_dt <- cbind( motif_footprints_dt[overlap_dt$queryHits,.(motif,motif_score)],\n",
    "                          promoter_dt[overlap_dt$subjectHits,.(tx_name,gene_type)] )\n",
    "overlap_motif_dt <- merge( overlap_motif_dt, cds_tx_dt[,.(TXNAME,SYMBOL)], by.x=\"tx_name\",\n",
    "                         by.y=\"TXNAME\" )[,!c(\"tx_name\")]\n",
    "rm(promoter_dt)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n",
      "Warning message in chisq.test(fisher_mat[1, ], p = num_null/sum(num_null)):\n",
      "“Chi-squared approximation may be incorrect”\n"
     ]
    }
   ],
   "source": [
    "motifs <- unique(overlap_motif_dt$motif)\n",
    "#           is edge     is not edge\n",
    "#has motif\n",
    "#does not have motif\n",
    "num_total_genes <- length(all_genes)\n",
    "p_value_dt <- data.table( motif=motifs, p_value_fisher=1, p_value_chi_sq=1 )\n",
    "frac_occ_dt <- data.table( motif=motifs, edge_frac=-1, non_edge_frac=-1 )\n",
    "num_edge_genes <- length(unique(overlap_motif_dt[gene_type == \"Edge\",SYMBOL]))\n",
    "num_non_edge_genes <- length(unique(overlap_motif_dt[gene_type != \"Edge\",SYMBOL]))\n",
    "\n",
    "for (motif_ in motifs) {\n",
    "    motif_dt <- overlap_motif_dt[motif == motif_,.(motif,gene_type,SYMBOL)] %>% unique(.)\n",
    "    num_occ_edge <- nrow( motif_dt[gene_type == \"Edge\",] )\n",
    "    num_occ_not_edge <- nrow( motif_dt[gene_type != \"Edge\",] )\n",
    "    num_not_occ_edge <- num_edge_genes - num_occ_edge\n",
    "    num_occ <- nrow(motif_dt)\n",
    "    \n",
    "    num_genes_without_motif <- overlap_motif_dt[motif != motif_ & gene_type != \"Edge\",]$SYMBOL %>% unique(.) %>% length(.)\n",
    "\n",
    "    num_not_occ_not_edge <- num_genes_without_motif-num_not_occ_edge\n",
    "    fisher_mat <- matrix(c(num_occ_edge,num_occ_not_edge,num_not_occ_edge,num_genes_without_motif),\n",
    "                        nrow=2,ncol=2,byrow=T)\n",
    "    fisher_p_value <- fisher.test( fisher_mat, alternative = \"g\" )$p.value\n",
    "    num_null <- c(num_not_occ_edge,num_not_occ_not_edge)\n",
    "    chisq_p_value <- chisq.test( fisher_mat[1,], p=num_null/sum(num_null) )$p.value\n",
    "\n",
    "    p_value_dt[motif==motif_,`:=`(p_value_fisher=fisher_p_value,p_value_chi_sq=chisq_p_value)]\n",
    "    \n",
    "    frac_occ_dt[motif==motif_,`:=`(edge_frac=num_occ_edge/num_edge_genes,\n",
    "                                   non_edge_frac=num_occ_not_edge/num_non_edge_genes)]\n",
    "}\n",
    "# p_value_dt[,p_adj:=p.adjust(p_value)]\n",
    "frac_occ_dt[,ratio:=edge_frac/non_edge_frac]\n",
    "frac_occ_dt <- merge( frac_occ_dt, p_value_dt )[ratio > 1,]\n",
    "frac_occ_dt[,p_adj_chi_sq:=p.adjust(p_value_chi_sq)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Edge cells in other datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_dir <- \"/Users/sreenivasagopv2/Data/sc-funnel/GSE81547/\"\n",
    "file_paths = list.files(gene_exp_dir,full.names=T)\n",
    "gene_exp_dt <- data.table()\n",
    "col_idx <- 1\n",
    "for (file_path in file_paths) {\n",
    "    if (grepl(\".gz\",file_path)) {\n",
    "        print(col_idx)\n",
    "        flush.console()\n",
    "        dt <- fread( file_path )\n",
    "        gsm_num <- str_match( file_path, \"GSM[0-9][0-9]*\" )[1]\n",
    "        if (nrow(gene_exp_dt) == 0 ) {\n",
    "            gene_exp_dt <- dt %>% setnames(.,\"V2\",gsm_num)\n",
    "        } else {\n",
    "            gene_exp_dt <- cbind( gene_exp_dt, dt[,list(V2)] ) %>% setnames(.,\"V2\",gsm_num)\n",
    "        }\n",
    "        col_idx <- col_idx + 1\n",
    "    }\n",
    "}\n",
    "out_path <- \"/Users/sreenivasagopv2/Data/sc-funnel/GSE81547/GSE81547_mat.tsv.gz\"\n",
    "fwrite(gene_exp_dt,out_path,quote=F,row.names=F,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_mat <- read_gene_exp_mat(out_path)\n",
    "aging_obj <- create_full_seurat_object( gene_exp_mat )\n",
    "aging_obj <- subset( aging_obj, features=rownames(seurat_obj[[\"RNA\"]]@counts) )\n",
    "\n",
    "rm(gene_exp_mat)\n",
    "\n",
    "geo_info <- getGEO(\"GSE81547\",getGPL=F)\n",
    "geo_info_df <- pData(geo_info$GSE81547_series_matrix.txt.gz)\n",
    "donor_age <- str_match(geo_info_df[,\"title\"],\"^[0-9][0-9]*\")\n",
    "info_to_add_df <- data.frame( age=as.integer(donor_age), row.names=Cells(aging_obj))\n",
    "aging_obj <- AddMetaData( aging_obj, info_to_add_df )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Finding acinar cells\n",
    "aging_obj <- NormalizeData( aging_obj ) %>% ScaleData(.) %>% FindVariableFeatures(.) %>% RunPCA(.,npcs=50) %>% FindNeighbors( . ) %>% FindClusters(.) %>% RunUMAP(.,dims=1:50)\n",
    "aging_meta_data_dt <- data.table( aging_obj@meta.data, keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "prss1_exp_dt <- data.table( FetchData( aging_obj, vars=c(\"PRSS1\",\"seurat_clusters\")) )\n",
    "print(prss1_exp_dt[,mean(PRSS1),by=seurat_clusters][order(-V1),]) #This gave clusters 1,13,15,16\n",
    "aging_acinar_cells <- aging_meta_data_dt[seurat_clusters %in% c(1,15,13,16),cell.name]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Signature-based evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_acinar_up_genes <- diff_exp_dt_list[[\"Acinar cell\"]][avg_logFC > 0,gene_name]\n",
    "length(edge_acinar_up_genes)\n",
    "aging_acinar_data_mat <- aging_obj[[\"RNA\"]]@data[,aging_acinar_cells]\n",
    "edge_acinar_up_genes <- edge_acinar_up_genes[edge_acinar_up_genes %in% rownames(aging_acinar_data_mat)]\n",
    "\n",
    "\n",
    "aging_cell_edge_scores <- Matrix::colMeans(aging_acinar_data_mat[edge_acinar_up_genes,])\n",
    "aging_edge_dt <- data.table( cell_name=names(aging_cell_edge_scores), edge_score=aging_cell_edge_scores )\n",
    "cells_donor_age <- aging_meta_data_dt[cell.name %in% acinar_cells,][order(match(cell.name,names(aging_cell_edge_scores))),age]\n",
    "aging_edge_dt[,donor_age:=cells_donor_age]\n",
    "\n",
    "options(repr.plot.width=5, repr.plot.height=20)\n",
    "aging_edge_dt$donor_age <- factor(aging_edge_dt$donor_age,levels=sort(unique(cells_donor_age)))\n",
    "ggplot( aging_edge_dt ) + geom_density(aes(x=edge_score)) + \n",
    "facet_wrap(~donor_age,ncol=1) + theme_gray(base_size=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=5)\n",
    "ggplot( aging_edge_dt ) + geom_boxplot(aes(x=donor_age,y=edge_score)) + theme_gray(base_size=20) +\n",
    "stat_compare_means(aes(x=donor_age,y=edge_score))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cosine similarity based evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_edge_acinar_genes <- diff_exp_dt_list[[\"Acinar cell\"]]$gene_name\n",
    "\n",
    "# all_edge_acinar_genes <- sample(rownames(aging_obj[[\"RNA\"]]@counts),size=1000)\n",
    "all_edge_acinar_genes <- all_edge_acinar_genes[all_edge_acinar_genes %in% rownames(aging_acinar_data_mat)]\n",
    "original_edge_acinar_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                                           cell_category == \"edge\",cell.name]\n",
    "original_non_edge_acinar_cells <- edge_info_all$edge_center_dt[normal_cell_type == \"Acinar cell\" & \n",
    "                                                           cell_category == \"center\",cell.name]\n",
    "original_acinar_cells <- c(original_edge_acinar_cells,original_non_edge_acinar_cells)\n",
    "original_acinar_obj <- subset( seurat_obj, idents=\"Acinar cell\")# %>% ScaleData(features=all_edge_acinar_genes)\n",
    "acinar_data_mat <- as.matrix(original_acinar_obj[[\"RNA\"]]@data[all_edge_acinar_genes,original_acinar_cells])\n",
    "aging_acinar_obj <- subset( aging_obj, cells=aging_acinar_cells )# %>% ScaleData(features=all_edge_acinar_genes)\n",
    "\n",
    "aging_acinar_data_mat <- aging_acinar_obj[[\"RNA\"]]@data[all_edge_acinar_genes,aging_acinar_cells]\n",
    "rm(aging_acinar_obj)\n",
    "rm(original_acinar_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cosine_matrix <- function(X,Y) {\n",
    "    norm_X <- sqrt(colSums(X^2))\n",
    "    norm_Y <- sqrt(colSums(Y^2))\n",
    "    \n",
    "    dimnames_X <- list(rownames(X),colnames(X))\n",
    "    dimnames_Y <- list(rownames(Y),colnames(Y))\n",
    "\n",
    "    X <- sapply(1:ncol(X),function(idx){return(X[,idx]/norm_X[idx])})\n",
    "    Y <- sapply(1:ncol(Y),function(idx){return(Y[,idx]/norm_Y[idx])})\n",
    "    \n",
    "    dimnames(X) <- dimnames_X\n",
    "    dimnames(Y) <- dimnames_Y\n",
    "    \n",
    "    dist_mat <- t(X) %*% Y\n",
    "    return(dist_mat)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_cos_sim_mat <- get_cosine_matrix(acinar_data_mat,acinar_data_mat)\n",
    "\n",
    "original_edge_medoid <- names(which.max(rowSums(original_cos_sim_mat[original_edge_acinar_cells,original_edge_acinar_cells])))\n",
    "original_non_edge_medoid <- which.max(rowSums(original_cos_sim_mat[original_non_edge_acinar_cells,\n",
    "                                                            original_non_edge_acinar_cells]))\n",
    "\n",
    "average_non_edge_cos_sim <- original_cos_sim_mat[original_non_edge_acinar_cells,\n",
    "                                                          original_edge_medoid]\n",
    "num_original_edge_cells <- length(original_edge_acinar_cells)\n",
    "average_edge_cos_sim <- original_cos_sim_mat[original_edge_acinar_cells,\n",
    "                                                          original_edge_medoid]# - (1/num_original_edge_cells)\n",
    "#average_edge_cos_sim <- ((num_original_edge_cells-1)/num_original_edge_cells) * average_edge_cos_sim\n",
    "\n",
    "average_cos_sim <- c(average_non_edge_cos_sim,average_edge_cos_sim)\n",
    "original_sim_stats_dt <- data.table( average_cos_sim=average_cos_sim, \n",
    "                                    cell_name=names(average_cos_sim), cell_category=\"edge\" )\n",
    "original_sim_stats_dt[cell_name %in% original_non_edge_acinar_cells,cell_category:=\"non-edge\"]\n",
    "\n",
    "aging_edge_cos_sim_mat <- get_cosine_matrix( aging_acinar_data_mat, acinar_data_mat )\n",
    "average_aging_sim_with_edge <- aging_edge_cos_sim_mat[aging_acinar_cells,original_edge_medoid]#original_edge_acinar_cells])\n",
    "average_aging_sim_with_non_edge <- aging_edge_cos_sim_mat[aging_acinar_cells,original_non_edge_medoid]\n",
    "\n",
    "aging_sim_stats_dt <- data.table( `Similarity with Edge`=average_aging_sim_with_edge, \n",
    "                                 `Similarity with Non-Edge`=average_aging_sim_with_non_edge,\n",
    "                                 cell_name=names(average_aging_sim_with_edge), age=-1 )\n",
    "cells_donor_age <- aging_meta_data_dt[cell.name %in% aging_acinar_cells,][order(match(cell.name,names(average_aging_sim_with_edge))),age]\n",
    "aging_sim_stats_dt[,age:=cells_donor_age]\n",
    "\n",
    "\n",
    "edge_kde <- kde( average_edge_cos_sim, eval.points = average_aging_sim_with_edge )\n",
    "non_edge_kde <- kde( average_non_edge_cos_sim, eval.points = average_aging_sim_with_non_edge )\n",
    "edge_probability <- edge_kde$estimate\n",
    "non_edge_probability <- non_edge_kde$estimate\n",
    "likelihood_ratio <- edge_probability/non_edge_probability\n",
    "likelihood_dt <- data.table( likelihood_ratio=likelihood_ratio, cell_name=names(average_aging_sim_with_edge))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "melted_dt <- melt(aging_sim_stats_dt,id.vars=c(\"cell_name\",\"age\"))\n",
    "\n",
    "ggplot( melted_dt[!grepl(\"Non-Edge\",variable),] ) + geom_density(aes(x=value,color=variable)) + facet_wrap(~age,ncol=4) +\n",
    "theme_classic(base_size=18) + theme(legend.position=\"none\",panel.grid.major=element_line()) + \n",
    "xlab(\"Cosine Similarity with Edge Acinar state\")\n",
    "\n",
    "ggboxplot( melted_dt[!grepl(\"Non-Edge\",variable),], x=\"age\", y=\"value\" ) +\n",
    "theme_classic(base_size=18) + theme(legend.position=\"none\",panel.grid.major=element_line()) + \n",
    "ylab(\"Cosine Similarity with Edge Acinar state\") + xlab(\"Donor Age\")\n",
    "\n",
    "4 + \"\"\n",
    "\n",
    "merged_stats_dt <- merge( aging_sim_stats_dt, likelihood_dt, by=\"cell_name\" )\n",
    "merged_stats_dt[,cell_call:=\"Non-Edge\"]\n",
    "merged_stats_dt[likelihood_ratio > 1.2,cell_call:=\"Edge\"]\n",
    "\n",
    "total_N_dt <- merged_stats_dt[,.N,by=\"age\"]\n",
    "merged_stats_dt <- merge( merged_stats_dt, total_N_dt, by=\"age\" ) %>% setnames(.,\"N\",\"N_total\")\n",
    "\n",
    "N_edge_dt <- merged_stats_dt[cell_call == \"Edge\",.N,by=\"age\"]  %>% setnames(.,\"N\",\"N_edge\")\n",
    "merged_stats_dt <- merge( merged_stats_dt, N_edge_dt, by=\"age\")\n",
    "\n",
    "# merged_stats_dt[,facet_title:=paste(\"Age:\",age,\"f=\",round(N_edge/N_total,2),paste0(\"(\",paste(N_edge,N_total,sep=\"/\"),\")\"))]\n",
    "# melted_dt <- melt(merged_stats_dt,id.vars=c(\"cell_name\",\"age\",\"likelihood_ratio\",\n",
    "#                                                \"cell_call\",\"N_total\",\"N_edge\",\"facet_title\"))\n",
    "\n",
    "# ggplot( melted_dt ) + geom_density(aes(x=value,color=variable)) + facet_wrap(~facet_title,ncol=4) +\n",
    "# theme_classic(base_size=18) + theme(legend.position=\"bottom\") + xlab(\"Cosine similarity\")\n",
    "#annotate(\"text\",x=0.3,y=20,label=facet_title)#+ xlim(-1,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=9, repr.plot.height=6)\n",
    "\n",
    "ggboxplot( melted_dt[!grepl(\"Non-Edge\",variable),], x=\"age\", y=\"value\",color=\"age\", add=\"jitter\",\n",
    "         size=1.0) +\n",
    "theme_classic(base_size=18) + theme(legend.position=\"none\",panel.grid.major=element_line()) + \n",
    "ylab(\"Cosine Similarity with Edge Acinar state\") + xlab(\"Donor Age\") + \n",
    "stat_compare_means(ref.group = \"1\",label=\"p.signif\",size=8,label.y=0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=8)\n",
    "ggplot( original_sim_stats_dt ) + geom_density(aes(x=average_cos_sim,color=cell_category)) +\n",
    "theme_gray(base_size=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( aging_sim_stats_dt ) + geom_point(aes(x=`Similarity with Non-Edge`,y=`Similarity with Edge`) ) +\n",
    "geom_line(aes(x=`Similarity with Non-Edge`,y=`Similarity with Non-Edge`)) + facet_wrap(~age,ncol=4) + \n",
    "theme_gray(base_size=18) #+ xlim(0,1) + ylim(0,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_rankings <- AUCell_buildRankings(aging_acinar_data_mat, nCores=6, plotStats=F)\n",
    "\n",
    "edge_gene_set = list(\"Edgeness\"=edge_acinar_up_genes)\n",
    "cells_AUC <- AUCell_calcAUC(edge_gene_set, cells_rankings,nCores=6 )\n",
    "pathway_scores_mat <- getAUC(cells_AUC)# %>% setnames(.,\"rn\",\"gene_set\") %>% melt(.,id.vars=\"gene_set\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aucell_aging_dt <- data.table( t(pathway_scores_mat), keep.rownames=T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "cells_donor_age <- aging_meta_data_dt[cell.name %in% aucell_aging_dt$cell.name,][order(match(cell.name,aucell_aging_dt$cell.name)),age]\n",
    "aucell_aging_dt[,age:=cells_donor_age]\n",
    "aucell_aging_dt <- aucell_aging_dt[order(age),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking which sample edge cells predominantly come from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "edge_info_all_merged <- merge( edge_info_all$edge_center_dt, meta_data_dt[,.(cell.name,sample)] )\n",
    "ordered_samples <- meta_data_dt[,.N,by=sample][order(-N)]$sample\n",
    "edge_info_all_merged$sample <- factor( edge_info_all_merged$sample, levels = ordered_samples)\n",
    "\n",
    "cell_types_present <- unique(edge_info_all$edge_center_dt$normal_cell_type)\n",
    "combined_fisher_dt <- data.table()\n",
    "for (cell_type_ in cell_types_present) {\n",
    "    temp_dt <- edge_info_all_merged[normal_cell_type == cell_type_,]\n",
    "    ordered_N_dt <- meta_data_dt[cluster == cell_type_,.N,by=sample][order(-N)]\n",
    "    ordered_N_edge_dt <- temp_dt[cell_category==\"edge\",.N,by=sample][,.(sample,N_edge=N)]\n",
    "    ordered_N_dt <- merge( ordered_N_dt, ordered_N_edge_dt, by=\"sample\" )[order(-N),]\n",
    "    \n",
    "    N_edge <- sum(ordered_N_dt$N_edge)\n",
    "    N_total <- sum(ordered_N_dt$N)\n",
    "    \n",
    "    ordered_samples <- ordered_N_dt$sample\n",
    "\n",
    "    p_values <- c()\n",
    "    odds_values <- c()\n",
    "    for (sample_ in ordered_samples) {\n",
    "        N_edge_from_sample <- ordered_N_dt[sample == sample_,N_edge]\n",
    "        N_non_edge_from_sample <- ordered_N_dt[sample == sample_,N] - N_edge_from_sample\n",
    "        N_edge_from_other_samples <- N_edge - N_edge_from_sample\n",
    "        N_non_edge_from_other_samples <- N_total - N_edge_from_sample - N_non_edge_from_sample - N_edge_from_other_samples\n",
    "        odds_mat <- matrix(c(N_edge_from_sample,N_non_edge_from_sample,\n",
    "                             N_edge_from_other_samples,N_non_edge_from_other_samples),byrow=T,nrow=2,ncol=2)\n",
    "        f_test <- fisher.test(odds_mat,alternative=\"greater\")\n",
    "        p_values <- c(p_values,f_test$p.value)\n",
    "        odds_values <- c(odds_values,f_test$estimate)\n",
    "    }\n",
    "    ordered_N_dt[,`:=`(p_value=p_values,normal_cell_type=cell_type_,odds_ratio=odds_values)]\n",
    "    combined_fisher_dt <- rbind( combined_fisher_dt, ordered_N_dt )\n",
    "  \n",
    "}\n",
    "combined_fisher_dt$adj_p_value <- p.adjust(combined_fisher_dt$p_value,method=\"bonferroni\")\n",
    "combined_fisher_dt[adj_p_value < 0.1,]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_info_list <- list()\n",
    "samples <- unique( meta_data_dt$sample )\n",
    "for (sample_id in samples) {\n",
    "    sample_seurat_obj <- subset( seurat_obj, cells=meta_data_dt[sample == sample_id,cell.name])\n",
    "    sample_seurat_obj <- NormalizeData(sample_seurat_obj) %>% FindVariableFeatures(.,nfeatures=1000) %>% \n",
    "    ScaleData(.) %>% RunPCA(.,npcs=50,verbose=F)  %>% RunUMAP(.,dims=1:50,verbose=F)\n",
    "    \n",
    "    homotypic.prop <- modelHomotypic(sample_seurat_obj@meta.data$cluster)\n",
    "    nExp_poi <- round(0.05*nrow(sample_seurat_obj@meta.data))\n",
    "    #nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop)) \n",
    "    \n",
    "    sample_param_sweep <- paramSweep_v3(sample_seurat_obj, PCs = 1:50, sct = FALSE)\n",
    "    sample_param_sweep_summary <- summarizeSweep(sample_param_sweep, GT = FALSE)\n",
    "    sweep_dt <- data.table( find.pK( sample_param_sweep_summary ) )\n",
    "    optimal_pK <- as.double(as.vector(sweep_dt[order(-BCmetric),][1]$pK))\n",
    "    sample_seurat_obj <- doubletFinder_v3(sample_seurat_obj, \n",
    "                               PCs = 1:50, pN = 0.25, pK = optimal_pK, nExp = nExp_poi, \n",
    "                                          reuse.pANN = FALSE, sct = FALSE)\n",
    "    \n",
    "    sample_meta_data_dt <- data.table( sample_seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "    doublet_class_col <- paste(\"DF.classifications_0.25\",optimal_pK,nExp_poi,sep=\"_\")\n",
    "    doublet_info_list[[sample_id]] <- sample_meta_data_dt[,c(\"cell.name\",doublet_class_col),with=F] %>%  setnames(.,doublet_class_col,\"doublet_class\")\n",
    "    #4 + \"\"\n",
    "    #meta_data_dt <- data.table( subset_seurat_obj@meta.data, keep.rownames = T ) %>% setnames(.,\"rn\",\"cell.name\")\n",
    "}\n",
    "\n",
    "merged_doublet_dt <- rbindlist(doublet_info_list)\n",
    "cell_doublets <- meta_data_dt[cell.name %in% merged_doublet_dt[doublet_class == \"Doublet\",cell.name],cell.name]\n",
    "fwrite( data.table(cell.name=cell_doublets), \"cell_doublets.tsv\", sep=\"\\t\", col.names=F, quote=F)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "324.2px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
